// @microsoft/teamsfx-api/build/qm/question.js
"use strict";
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT license.
Object.defineProperty(exports, "__esModule", { value: true });
exports.QTreeNode = void 0;
/**
 * QTreeNode is the tree node data structure, which have three main properties:
 * - data: data is either a group or question. Questions can be organized into a group, which has the same trigger condition.
 * - condition: trigger condition for this node to be activated;
 * - children: child questions that will be activated according their trigger condition.
 */
class QTreeNode {
    constructor(data) {
        this.data = data;
    }
    addChild(node) {
        if (!this.children) {
            this.children = [];
        }
        this.children.push(node);
        return this;
    }
    validate() {
        //1. validate the cycle dependency
        //2. validate the name uniqueness
        //3. validate the params of RPC
        // if (this.data.type === NodeType.group && (!this.children || this.children.length === 0)) return false;
        return true;
    }
    /**
     * trim the tree
     */
    trim() {
        if (this.children) {
            const newChildren = [];
            for (const node of this.children) {
                const trimmed = node.trim();
                if (trimmed)
                    newChildren.push(trimmed);
            }
            this.children = newChildren;
        }
        if (this.data.type === "group") {
            if (!this.children || this.children.length === 0)
                return undefined;
            if (this.children.length === 1) {
                const child = this.children[0];
                if (!(this.condition && child.condition)) {
                    child.condition || (child.condition = this.condition);
                    return child;
                }
            }
        }
        return this;
    }
}
exports.QTreeNode = QTreeNode;
