// @microsoft/teamsfx-core/build/component/driver/teamsApp/configure.js
"use strict";
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT license.
Object.defineProperty(exports, "__esModule", { value: true });
exports.ConfigureTeamsAppDriver = exports.outputNames = exports.actionName = void 0;
const tslib_1 = require("tslib");
const teamsfx_api_1 = require("@microsoft/teamsfx-api");
const fs_extra_1 = tslib_1.__importDefault(require("fs-extra"));
const lib_1 = require("@feathersjs/hooks/lib");
const wrapUtil_1 = require("../util/wrapUtil");
const addStartAndEndTelemetry_1 = require("../middleware/addStartAndEndTelemetry");
const appStudioClient_1 = require("../../resource/appManifest/appStudioClient");
const results_1 = require("../../resource/appManifest/results");
const telemetry_1 = require("../../resource/appManifest/utils/telemetry");
const errors_1 = require("../../resource/appManifest/errors");
const tools_1 = require("../../../common/tools");
const localizeUtils_1 = require("../../../common/localizeUtils");
const typedi_1 = require("typedi");
const common_1 = require("../../utils/common");
exports.actionName = "teamsApp/update";
exports.outputNames = {
    TEAMS_APP_ID: "TEAMS_APP_ID",
    TEAMS_APP_TENANT_ID: "TEAMS_APP_TENANT_ID",
    TEAMS_APP_UPDATE_TIME: "TEAMS_APP_UPDATE_TIME",
};
let ConfigureTeamsAppDriver = class ConfigureTeamsAppDriver {
    constructor() {
        this.description = localizeUtils_1.getLocalizedString("driver.teamsApp.description.updateDriver");
    }
    async run(args, context) {
        const wrapContext = new wrapUtil_1.WrapDriverContext(context, exports.actionName, exports.actionName);
        const res = await this.update(args, wrapContext);
        return res;
    }
    async execute(args, context) {
        const wrapContext = new wrapUtil_1.WrapDriverContext(context, exports.actionName, exports.actionName);
        const res = await this.update(args, wrapContext);
        return {
            result: res,
            summaries: wrapContext.summaries,
        };
    }
    async update(args, context) {
        var _a, _b;
        telemetry_1.TelemetryUtils.init(context);
        const result = this.validateArgs(args);
        if (result.isErr()) {
            return teamsfx_api_1.err(result.error);
        }
        const appStudioTokenRes = await context.m365TokenProvider.getAccessToken({
            scopes: tools_1.AppStudioScopes,
        });
        if (appStudioTokenRes.isErr()) {
            return teamsfx_api_1.err(appStudioTokenRes.error);
        }
        const appStudioToken = appStudioTokenRes.value;
        const appPackagePath = common_1.getAbsolutePath(args.appPackagePath, context.projectPath);
        if (!(await fs_extra_1.default.pathExists(appPackagePath))) {
            return teamsfx_api_1.err(results_1.AppStudioResultFactory.UserError(errors_1.AppStudioError.FileNotFoundError.name, errors_1.AppStudioError.FileNotFoundError.message(args.appPackagePath), "https://aka.ms/teamsfx-actions/teamsapp-update"));
        }
        const archivedFile = await fs_extra_1.default.readFile(appPackagePath);
        const progressHandler = (_a = context.ui) === null || _a === void 0 ? void 0 : _a.createProgressBar(localizeUtils_1.getLocalizedString("driver.teamsApp.progressBar.updateTeamsAppTitle"), 1);
        progressHandler === null || progressHandler === void 0 ? void 0 : progressHandler.start();
        try {
            let message = localizeUtils_1.getLocalizedString("driver.teamsApp.progressBar.updateTeamsAppStepMessage");
            progressHandler === null || progressHandler === void 0 ? void 0 : progressHandler.next(message);
            context.addSummary(message);
            const appDefinition = await appStudioClient_1.AppStudioClient.importApp(archivedFile, appStudioToken, context.logProvider, true);
            message = localizeUtils_1.getLocalizedString("plugins.appstudio.teamsAppUpdatedLog", appDefinition.teamsAppId);
            context.logProvider.info(message);
            context.addSummary(message);
            if (context.platform === teamsfx_api_1.Platform.VSCode) {
                (_b = context.ui) === null || _b === void 0 ? void 0 : _b.showMessage("info", message, false);
            }
            progressHandler === null || progressHandler === void 0 ? void 0 : progressHandler.end(true);
            return teamsfx_api_1.ok(new Map([
                [exports.outputNames.TEAMS_APP_ID, appDefinition.teamsAppId],
                [exports.outputNames.TEAMS_APP_TENANT_ID, appDefinition.tenantId],
                [exports.outputNames.TEAMS_APP_UPDATE_TIME, appDefinition.updatedAt],
            ]));
        }
        catch (e) {
            progressHandler === null || progressHandler === void 0 ? void 0 : progressHandler.end(false);
            return teamsfx_api_1.err(results_1.AppStudioResultFactory.SystemError(errors_1.AppStudioError.TeamsAppUpdateFailedError.name, errors_1.AppStudioError.TeamsAppUpdateFailedError.message(e), "https://aka.ms/teamsfx-actions/teamsapp-update"));
        }
    }
    validateArgs(args) {
        const invalidParams = [];
        if (!args || !args.appPackagePath) {
            invalidParams.push("appPackagePath");
        }
        if (invalidParams.length > 0) {
            return teamsfx_api_1.err(results_1.AppStudioResultFactory.UserError(errors_1.AppStudioError.InvalidParameterError.name, errors_1.AppStudioError.InvalidParameterError.message(exports.actionName, invalidParams), "https://aka.ms/teamsfx-actions/teamsapp-update"));
        }
        else {
            return teamsfx_api_1.ok(undefined);
        }
    }
};
tslib_1.__decorate([
    lib_1.hooks([addStartAndEndTelemetry_1.addStartAndEndTelemetry(exports.actionName, exports.actionName)]),
    tslib_1.__metadata("design:type", Function),
    tslib_1.__metadata("design:paramtypes", [Object, wrapUtil_1.WrapDriverContext]),
    tslib_1.__metadata("design:returntype", Promise)
], ConfigureTeamsAppDriver.prototype, "update", null);
ConfigureTeamsAppDriver = tslib_1.__decorate([
    typedi_1.Service(exports.actionName)
], ConfigureTeamsAppDriver);
exports.ConfigureTeamsAppDriver = ConfigureTeamsAppDriver;
