// @microsoft/teamsfx-core/build/component/feature/cicd/constants.js
"use strict";
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT license.
Object.defineProperty(exports, "__esModule", { value: true });
exports.Suggestions = exports.PluginCICD = exports.PluginSolution = exports.TelemetryValues = exports.TelemetryKeys = exports.Alias = exports.ErrorNames = exports.LifecycleFuncNames = void 0;
const localizeUtils_1 = require("../../../common/localizeUtils");
class LifecycleFuncNames {
}
exports.LifecycleFuncNames = LifecycleFuncNames;
LifecycleFuncNames.ADD_CICD_WORKFLOWS = "add-cicd-workflows";
class ErrorNames {
}
exports.ErrorNames = ErrorNames;
ErrorNames.INTERNAL_ERROR = "InternalError";
ErrorNames.NO_PROJECT_OPENED_ERROR = "NoProjectOpenedError";
ErrorNames.FILE_SYSTEM_ERROR = "FileSystemError";
class Alias {
}
exports.Alias = Alias;
Alias.TEAMS_CICD_PLUGIN = "CICD";
Alias.TEAMS_FX = "Teamsfx";
class TelemetryKeys {
}
exports.TelemetryKeys = TelemetryKeys;
TelemetryKeys.Component = "component";
TelemetryKeys.Success = "success";
TelemetryKeys.ErrorType = "error-type";
TelemetryKeys.ErrorMessage = "error-message";
TelemetryKeys.ErrorCode = "error-code";
TelemetryKeys.AppId = "appid";
class TelemetryValues {
}
exports.TelemetryValues = TelemetryValues;
TelemetryValues.Success = "yes";
TelemetryValues.Fail = "no";
TelemetryValues.UserError = "user";
TelemetryValues.SystemError = "system";
class PluginSolution {
}
exports.PluginSolution = PluginSolution;
PluginSolution.PLUGIN_NAME = "solution";
PluginSolution.REMOTE_TEAMS_APPID = "remoteTeamsAppId";
class PluginCICD {
}
exports.PluginCICD = PluginCICD;
PluginCICD.PLUGIN_NAME = "fx-resource-cicd";
class Suggestions {
}
exports.Suggestions = Suggestions;
Suggestions.RETRY_THE_CURRENT_STEP = () => localizeUtils_1.getLocalizedString("suggestions.retryTheCurrentStep");
Suggestions.CREATE_PROJECT_OR_OPEN_EXISTING = () => localizeUtils_1.getLocalizedString("plugins.cicd.suggestions.createProjectOrOpenExisting");
Suggestions.CHECK_PERMISSION = () => localizeUtils_1.getLocalizedString("plugins.cicd.suggestions.checkPermission");
