// @microsoft/teamsfx-core/build/component/feature/cicd/questions.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.questionNames = exports.providerIdToLabel = exports.templateIdToLabel = exports.templateOptions = exports.publishOption = exports.provisionOption = exports.cdOption = exports.ciOption = exports.providerOptions = exports.jenkinsOption = exports.azdoOption = exports.githubOption = void 0;
const localizeUtils_1 = require("../../../common/localizeUtils");
exports.githubOption = {
    id: "github",
    label: "GitHub",
    detail: "",
};
exports.azdoOption = {
    id: "azdo",
    label: "Azure DevOps",
    detail: "",
};
exports.jenkinsOption = {
    id: "jenkins",
    label: "Jenkins",
    detail: "",
};
exports.providerOptions = [exports.githubOption, exports.azdoOption, exports.jenkinsOption];
function ciOption() {
    return {
        id: "ci",
        label: "CI",
        detail: localizeUtils_1.getLocalizedString("plugins.cicd.ciOption.detail"),
    };
}
exports.ciOption = ciOption;
function cdOption() {
    return {
        id: "cd",
        label: "CD",
        detail: localizeUtils_1.getLocalizedString("plugins.cicd.cdOption.detail"),
    };
}
exports.cdOption = cdOption;
function provisionOption() {
    return {
        id: "provision",
        label: "Provision",
        detail: localizeUtils_1.getLocalizedString("plugins.cicd.provisionOption.detail"),
    };
}
exports.provisionOption = provisionOption;
function publishOption() {
    return {
        id: "publish",
        label: "Publish to Teams",
        detail: localizeUtils_1.getLocalizedString("plugins.cicd.publishOption.detail"),
    };
}
exports.publishOption = publishOption;
const templateOptions = () => [ciOption(), provisionOption(), cdOption(), publishOption()];
exports.templateOptions = templateOptions;
const templateIdLabelMap = new Map([
    [ciOption().id, ciOption().label],
    [cdOption().id, cdOption().label],
    [provisionOption().id, provisionOption().label],
    [publishOption().id, publishOption().label],
]);
const providerIdLabelMap = new Map([
    [exports.githubOption.id, exports.githubOption.label],
    [exports.azdoOption.id, exports.azdoOption.label],
    [exports.jenkinsOption.id, exports.jenkinsOption.label],
]);
function templateIdToLabel(templateId) {
    var _a;
    return (_a = templateIdLabelMap.get(templateId)) !== null && _a !== void 0 ? _a : templateId;
}
exports.templateIdToLabel = templateIdToLabel;
function providerIdToLabel(providerId) {
    var _a;
    return (_a = providerIdLabelMap.get(providerId)) !== null && _a !== void 0 ? _a : providerId;
}
exports.providerIdToLabel = providerIdToLabel;
var questionNames;
(function (questionNames) {
    questionNames["Provider"] = "provider";
    questionNames["Template"] = "template";
    questionNames["Environment"] = "target-env";
})(questionNames = exports.questionNames || (exports.questionNames = {}));
