// @microsoft/teamsfx-core/build/component/feature/cicd/result.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.FxCICDPluginResultFactory = void 0;
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT license.
const teamsfx_api_1 = require("@microsoft/teamsfx-api");
const constants_1 = require("./constants");
class FxCICDPluginResultFactory {
    static UserError(errorName, errorMessage, showHelpLink, innerError) {
        return teamsfx_api_1.err(new teamsfx_api_1.UserError({
            name: errorName,
            message: errorMessage[0],
            displayMessage: errorMessage[1],
            source: FxCICDPluginResultFactory.source,
            helpLink: showHelpLink ? FxCICDPluginResultFactory.defaultHelpLink : undefined,
            error: innerError,
        }));
    }
    static SystemError(errorName, errorMessage, innerError) {
        return teamsfx_api_1.err(new teamsfx_api_1.SystemError({
            name: errorName,
            message: errorMessage[0],
            displayMessage: errorMessage[1],
            source: FxCICDPluginResultFactory.source,
            issueLink: FxCICDPluginResultFactory.defaultIssueLink,
            error: innerError,
        }));
    }
    static Success(result) {
        return teamsfx_api_1.ok(result);
    }
}
exports.FxCICDPluginResultFactory = FxCICDPluginResultFactory;
FxCICDPluginResultFactory.source = constants_1.Alias.TEAMS_CICD_PLUGIN;
FxCICDPluginResultFactory.defaultHelpLink = "";
FxCICDPluginResultFactory.defaultIssueLink = "";
