// @microsoft/teamsfx-core/build/component/resource/aadApp/aadAppClient.js
"use strict";
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT license.
Object.defineProperty(exports, "__esModule", { value: true });
exports.AadAppClient = void 0;
const constants_1 = require("./constants");
const errorCodes_1 = require("./errorCodes");
const errors_1 = require("./errors");
const results_1 = require("./results");
const configs_1 = require("./utils/configs");
const dialog_1 = require("./utils/dialog");
const telemetry_1 = require("./utils/telemetry");
const tokenProvider_1 = require("./utils/tokenProvider");
const tools_1 = require("../../../common/tools");
const globalVars_1 = require("../../../core/globalVars");
const aadAppManifestManager_1 = require("./aadAppManifestManager");
const uuid_1 = require("uuid");
const graph_1 = require("./graph");
const appStudio_1 = require("./appStudio");
function delay(ms) {
    // tslint:disable-next-line no-string-based-set-timeout
    return new Promise((resolve) => setTimeout(resolve, ms));
}
class AadAppClient {
    static async createAadAppUsingManifest(stage, manifest, config) {
        var _a;
        try {
            manifest = (await this.retryHanlder(stage, () => aadAppManifestManager_1.AadAppManifestManager.createAadApp(tokenProvider_1.TokenProvider.token, manifest)));
            config.clientId = manifest.appId;
            config.objectId = manifest.id;
        }
        catch (error) {
            if (((_a = error === null || error === void 0 ? void 0 : error.response) === null || _a === void 0 ? void 0 : _a.status) == constants_1.Constants.statusCodeForbidden) {
                throw results_1.ResultFactory.UserError(errors_1.CreateAppForbiddenError.name, errors_1.CreateAppForbiddenError.message(), error, undefined, errors_1.CreateAppForbiddenError.helpLink);
            }
            else {
                throw AadAppClient.handleError(error, errors_1.CreateAppError);
            }
        }
    }
    static async createAadApp(stage, config) {
        try {
            const provisionObject = AadAppClient.getAadAppProvisionObject(config.displayName, config.oauth2PermissionScopeId);
            let provisionAadResponse;
            if (tokenProvider_1.TokenProvider.audience === tokenProvider_1.TokenAudience.AppStudio) {
                provisionAadResponse = (await this.retryHanlder(stage, () => appStudio_1.AppStudio.createAADAppV2(tokenProvider_1.TokenProvider.token, provisionObject)));
            }
            else {
                provisionAadResponse = (await this.retryHanlder(stage, () => graph_1.GraphClient.createAADApp(tokenProvider_1.TokenProvider.token, provisionObject)));
            }
            config.clientId = provisionAadResponse.appId;
            config.objectId = provisionAadResponse.id;
        }
        catch (error) {
            throw AadAppClient.handleError(error, errors_1.CreateAppError);
        }
    }
    static async createAadAppSecret(stage, config) {
        try {
            let createSecretObject;
            if (tokenProvider_1.TokenProvider.audience === tokenProvider_1.TokenAudience.AppStudio) {
                createSecretObject = (await AadAppClient.retryHanlder(stage, () => appStudio_1.AppStudio.createAADAppPassword(tokenProvider_1.TokenProvider.token, config.objectId)));
            }
            else {
                createSecretObject = (await AadAppClient.retryHanlder(stage, () => graph_1.GraphClient.createAadAppSecret(tokenProvider_1.TokenProvider.token, config.objectId)));
            }
            config.password = createSecretObject.value;
        }
        catch (error) {
            throw AadAppClient.handleError(error, errors_1.CreateSecretError);
        }
    }
    static async updateAadAppUsingManifest(stage, manifest, skip = false) {
        var _a;
        try {
            await AadAppClient.retryHanlder(stage, async () => {
                const preAuthorizedApplications = manifest.preAuthorizedApplications;
                manifest.preAuthorizedApplications = [];
                await aadAppManifestManager_1.AadAppManifestManager.updateAadApp(tokenProvider_1.TokenProvider.token, manifest);
                manifest.preAuthorizedApplications = preAuthorizedApplications;
                await aadAppManifestManager_1.AadAppManifestManager.updateAadApp(tokenProvider_1.TokenProvider.token, manifest);
            });
        }
        catch (error) {
            if (skip) {
                const message = constants_1.Messages.StepFailedAndSkipped(constants_1.ProgressDetail.UpdateAadApp, constants_1.Messages.UpdateAadHelpMessage());
                (_a = globalVars_1.TOOLS.logProvider) === null || _a === void 0 ? void 0 : _a.warning(constants_1.Messages.getLog(message));
                dialog_1.DialogUtils.show(message, constants_1.UILevels.Warn);
            }
            else {
                throw AadAppClient.handleError(error, tools_1.isAadManifestEnabled() ? errors_1.UpdateAadAppUsingManifestError : errors_1.UpdateAadAppError, error.message);
            }
        }
    }
    static async updateAadAppRedirectUri(stage, objectId, redirectUris, skip = false) {
        var _a;
        try {
            if (tokenProvider_1.TokenProvider.audience === tokenProvider_1.TokenAudience.AppStudio) {
                await AadAppClient.retryHanlder(stage, () => appStudio_1.AppStudio.updateAADApp(tokenProvider_1.TokenProvider.token, objectId, redirectUris));
            }
            else {
                await AadAppClient.retryHanlder(stage, () => graph_1.GraphClient.updateAADApp(tokenProvider_1.TokenProvider.token, objectId, redirectUris));
            }
        }
        catch (error) {
            if (skip) {
                const message = constants_1.Messages.StepFailedAndSkipped(constants_1.ProgressDetail.UpdateRedirectUri, constants_1.Messages.UpdateRedirectUriHelpMessage(configs_1.Utils.parseRedirectUriMessage(redirectUris)));
                (_a = globalVars_1.TOOLS.logProvider) === null || _a === void 0 ? void 0 : _a.warning(constants_1.Messages.getLog(message));
                dialog_1.DialogUtils.show(message, constants_1.UILevels.Warn);
            }
            else {
                throw AadAppClient.handleError(error, errors_1.UpdateRedirectUriError);
            }
        }
    }
    static async updateAadAppIdUri(stage, objectId, applicationIdUri, skip = false) {
        var _a;
        try {
            const updateAppIdObject = AadAppClient.getAadApplicationIdObject(applicationIdUri);
            if (tokenProvider_1.TokenProvider.audience === tokenProvider_1.TokenAudience.AppStudio) {
                await AadAppClient.retryHanlder(stage, () => appStudio_1.AppStudio.updateAADApp(tokenProvider_1.TokenProvider.token, objectId, updateAppIdObject));
            }
            else {
                await AadAppClient.retryHanlder(stage, () => graph_1.GraphClient.updateAADApp(tokenProvider_1.TokenProvider.token, objectId, updateAppIdObject));
            }
        }
        catch (error) {
            if (skip) {
                const message = constants_1.Messages.StepFailedAndSkipped(constants_1.ProgressDetail.UpdateAppIdUri, constants_1.Messages.UpdateAppIdUriHelpMessage(applicationIdUri));
                (_a = globalVars_1.TOOLS.logProvider) === null || _a === void 0 ? void 0 : _a.warning(constants_1.Messages.getLog(message));
                dialog_1.DialogUtils.show(message, constants_1.UILevels.Warn);
            }
            else {
                throw AadAppClient.handleError(error, errors_1.UpdateAppIdUriError);
            }
        }
    }
    static async updateAadAppPermission(stage, objectId, permissions, skip = false) {
        var _a;
        try {
            const updatePermissionObject = AadAppClient.getAadPermissionObject(permissions);
            if (tokenProvider_1.TokenProvider.audience === tokenProvider_1.TokenAudience.AppStudio) {
                await AadAppClient.retryHanlder(stage, () => appStudio_1.AppStudio.updateAADApp(tokenProvider_1.TokenProvider.token, objectId, updatePermissionObject));
            }
            else {
                await AadAppClient.retryHanlder(stage, () => graph_1.GraphClient.updateAADApp(tokenProvider_1.TokenProvider.token, objectId, updatePermissionObject));
            }
        }
        catch (error) {
            if (skip) {
                const message = constants_1.Messages.StepFailedAndSkipped(constants_1.ProgressDetail.UpdatePermission, constants_1.Messages.UpdatePermissionHelpMessage);
                (_a = globalVars_1.TOOLS.logProvider) === null || _a === void 0 ? void 0 : _a.warning(constants_1.Messages.getLog(message));
                dialog_1.DialogUtils.show(message, constants_1.UILevels.Warn);
            }
            else {
                throw AadAppClient.handleError(error, errors_1.UpdatePermissionError);
            }
        }
    }
    static async getAadAppUsingManifest(stage, objectId, clientSecret, oauth2PermissionScopeId, m365TokenProvider, envName) {
        var _a;
        let manifest;
        try {
            manifest = (await this.retryHanlder(stage, () => aadAppManifestManager_1.AadAppManifestManager.getAadAppManifest(tokenProvider_1.TokenProvider.token, objectId)));
        }
        catch (error) {
            const tenantId = await configs_1.Utils.getCurrentTenantId(m365TokenProvider);
            const fileName = configs_1.Utils.getConfigFileName(envName);
            throw AadAppClient.handleError(error, errors_1.GetAppError, objectId, tenantId, fileName);
        }
        const config = new configs_1.ProvisionConfig(!envName, false);
        // Check whether remote aad app contains scope id
        (_a = manifest.oauth2Permissions) === null || _a === void 0 ? void 0 : _a.forEach((oauth2Permission) => {
            if (oauth2Permission.value === "access_as_user") {
                config.oauth2PermissionScopeId = oauth2Permission.id;
            }
        });
        // If remote aad app doesn't contain scope id, use scope id in state file or create a new one
        if (!config.oauth2PermissionScopeId) {
            config.oauth2PermissionScopeId = oauth2PermissionScopeId ? oauth2PermissionScopeId : uuid_1.v4();
        }
        config.objectId = objectId;
        config.clientId = manifest.appId;
        config.password = clientSecret;
        return config;
    }
    static async getAadApp(stage, objectId, clientSecret, m365TokenProvider, envName, skip = false) {
        var _a, _b, _c, _d;
        let getAppObject;
        try {
            if (tokenProvider_1.TokenProvider.audience === tokenProvider_1.TokenAudience.AppStudio) {
                getAppObject = (await this.retryHanlder(stage, () => appStudio_1.AppStudio.getAadApp(tokenProvider_1.TokenProvider.token, objectId)));
            }
            else {
                getAppObject = (await this.retryHanlder(stage, () => graph_1.GraphClient.getAadApp(tokenProvider_1.TokenProvider.token, objectId)));
            }
        }
        catch (error) {
            const tenantId = await configs_1.Utils.getCurrentTenantId(m365TokenProvider);
            const fileName = configs_1.Utils.getConfigFileName(envName);
            throw AadAppClient.handleError(error, errors_1.GetAppError, objectId, tenantId, fileName);
        }
        const config = new configs_1.ProvisionConfig(!envName);
        if (((_a = getAppObject.api) === null || _a === void 0 ? void 0 : _a.oauth2PermissionScopes) &&
            ((_b = getAppObject.api) === null || _b === void 0 ? void 0 : _b.oauth2PermissionScopes[0]) &&
            ((_c = getAppObject.api) === null || _c === void 0 ? void 0 : _c.oauth2PermissionScopes[0].id)) {
            config.oauth2PermissionScopeId = (_d = getAppObject.api) === null || _d === void 0 ? void 0 : _d.oauth2PermissionScopes[0].id;
        }
        else {
            const fileName = configs_1.Utils.getConfigFileName(envName);
            throw results_1.ResultFactory.UserError(errors_1.GetAppConfigError.name, errors_1.GetAppConfigError.message(constants_1.ConfigKeys.oauth2PermissionScopeId, fileName));
        }
        config.objectId = objectId;
        config.clientId = getAppObject.appId;
        config.password = clientSecret;
        return config;
    }
    static async checkPermission(stage, objectId, userObjectId) {
        try {
            return (await this.retryHanlder(stage, () => graph_1.GraphClient.checkPermission(tokenProvider_1.TokenProvider.token, objectId, userObjectId)));
        }
        catch (error) {
            // TODO: Give out detailed help message for different errors.
            throw AadAppClient.handleError(error, errors_1.CheckPermissionError);
        }
    }
    static async grantPermission(ctx, objectId, userObjectId) {
        var _a, _b, _c;
        try {
            await graph_1.GraphClient.grantPermission(tokenProvider_1.TokenProvider.token, objectId, userObjectId);
        }
        catch (error) {
            if (((_b = (_a = error === null || error === void 0 ? void 0 : error.response) === null || _a === void 0 ? void 0 : _a.data) === null || _b === void 0 ? void 0 : _b.error.message) == constants_1.Constants.createOwnerDuplicatedMessage) {
                (_c = ctx.logProvider) === null || _c === void 0 ? void 0 : _c.info(constants_1.Messages.OwnerAlreadyAdded(userObjectId, objectId));
                return;
            }
            // TODO: Give out detailed help message for different errors.
            throw AadAppClient.handleError(error, errors_1.GrantPermissionError, constants_1.Constants.permissions.name, objectId);
        }
    }
    static async listCollaborator(stage, objectId) {
        try {
            return await this.retryHanlder(stage, () => graph_1.GraphClient.getAadOwners(tokenProvider_1.TokenProvider.token, objectId));
        }
        catch (error) {
            // TODO: Give out detailed help message for different errors.
            throw AadAppClient.handleError(error, errors_1.ListCollaboratorError);
        }
    }
    static async retryHanlder(stage, fn) {
        let retries = constants_1.Constants.maxRetryTimes;
        let response;
        while (retries > 0) {
            retries = retries - 1;
            try {
                response = await fn();
                telemetry_1.TelemetryUtils.sendEvent(stage, {
                    [constants_1.Telemetry.methodName]: fn.toString(),
                    [constants_1.Telemetry.retryTimes]: (constants_1.Constants.maxRetryTimes - retries - 1).toString(),
                });
                return response;
            }
            catch (error) {
                if (retries === 0) {
                    throw error;
                }
                else {
                    await delay(5000);
                }
            }
        }
        throw new Error(errors_1.AppStudioErrorMessage.ReachRetryLimit[0]);
    }
    static getAadAppProvisionObject(displayName, oauth2PermissionScopeId) {
        return {
            displayName: displayName,
            signInAudience: "AzureADMyOrg",
            api: {
                requestedAccessTokenVersion: 2,
                oauth2PermissionScopes: [
                    {
                        adminConsentDescription: "Allows Teams to call the app’s web APIs as the current user.",
                        adminConsentDisplayName: "Teams can access app’s web APIs",
                        id: oauth2PermissionScopeId,
                        isEnabled: true,
                        type: "User",
                        userConsentDescription: "Enable Teams to call this app’s web APIs with the same rights that you have",
                        userConsentDisplayName: "Teams can access app’s web APIs and make requests on your behalf",
                        value: "access_as_user",
                    },
                ],
                preAuthorizedApplications: tools_1.getAllowedAppIds().map((appId) => {
                    return {
                        appId,
                        delegatedPermissionIds: [oauth2PermissionScopeId],
                    };
                }),
            },
            optionalClaims: {
                accessToken: [
                    {
                        name: "idtyp",
                        essential: false,
                        additionalProperties: [],
                    },
                ],
            },
        };
    }
    static handleError(error, errorDetail, ...args) {
        var _a, _b, _c, _d, _e;
        if (((_a = error === null || error === void 0 ? void 0 : error.response) === null || _a === void 0 ? void 0 : _a.status) >= constants_1.Constants.statusCodeUserError &&
            ((_b = error === null || error === void 0 ? void 0 : error.response) === null || _b === void 0 ? void 0 : _b.status) < constants_1.Constants.statusCodeServerError) {
            // User Error
            // If known error code, will update help link.
            const errorCode = (_e = (_d = (_c = error === null || error === void 0 ? void 0 : error.response) === null || _c === void 0 ? void 0 : _c.data) === null || _d === void 0 ? void 0 : _d.error) === null || _e === void 0 ? void 0 : _e.code;
            const helpLink = errorCodes_1.GraphErrorCodes.get(errorCode);
            return results_1.ResultFactory.UserError(errorDetail.name, errorDetail.message(...args), error, undefined, helpLink !== null && helpLink !== void 0 ? helpLink : errorDetail.helpLink);
        }
        else {
            // System Error
            return results_1.ResultFactory.SystemError(errorDetail.name, errorDetail.message(...args), error);
        }
    }
    static getAadUrlObject(redirectUris) {
        return {
            web: {
                redirectUris: redirectUris,
            },
        };
    }
    static getAadApplicationIdObject(applicationIdUri) {
        return {
            identifierUris: [applicationIdUri],
        };
    }
    static getAadPermissionObject(permissions) {
        return {
            requiredResourceAccess: permissions,
        };
    }
}
exports.AadAppClient = AadAppClient;
