// @microsoft/teamsfx-core/build/common/local/projectSettingsHelper.js
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT license.
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ProjectSettingsHelper = void 0;
const constants_1 = require("../../component/constants");
const workflow_1 = require("../../component/workflow");
const projectSettingsHelperV3_1 = require("../projectSettingsHelperV3");
class ProjectSettingsHelper {
    static includeFrontend(projectSettings) {
        return projectSettingsHelperV3_1.hasAzureTab(projectSettings);
    }
    static includeBackend(projectSettings) {
        return projectSettingsHelperV3_1.hasApi(projectSettings);
    }
    static includeFuncHostedBot(projectSettings) {
        return projectSettingsHelperV3_1.hasFunctionBot(projectSettings);
    }
    static includeBot(projectSettings) {
        return projectSettingsHelperV3_1.hasBot(projectSettings);
    }
    static getBotCapabilities(projectSettings) {
        const bot = workflow_1.getComponent(projectSettings, constants_1.ComponentNames.TeamsBot);
        return (bot === null || bot === void 0 ? void 0 : bot.capabilities) || [];
    }
}
exports.ProjectSettingsHelper = ProjectSettingsHelper;
// keep the same logic as plugin.activate()
ProjectSettingsHelper.isSpfx = (projectSettings) => projectSettingsHelperV3_1.hasSPFxTab(projectSettings);
ProjectSettingsHelper.includeAAD = (projectSettings) => projectSettingsHelperV3_1.hasAAD(projectSettings);
ProjectSettingsHelper.includeSimpleAuth = (projectSettings) => projectSettingsHelperV3_1.hasSimpleAuth(projectSettings);
