// @microsoft/teamsfx-core/build/component/driver/aad/error/missingEnvInFileError.js
"use strict";
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT license.
Object.defineProperty(exports, "__esModule", { value: true });
exports.MissingEnvInFileUserError = void 0;
const teamsfx_api_1 = require("@microsoft/teamsfx-api");
const localizeUtils_1 = require("../../../../common/localizeUtils");
const errorCode = "MissingEnvironmentVariable";
const messageKey = "driver.aadApp.error.missingEnvInFile";
class MissingEnvInFileUserError extends teamsfx_api_1.UserError {
    constructor(actionName, missingEnvs, helpLink, additionalMessageKey, filePath) {
        let envList;
        if (Array.isArray(missingEnvs)) {
            envList = missingEnvs.join(", ");
        }
        else {
            envList = missingEnvs;
        }
        super({
            source: actionName,
            name: errorCode,
            message: localizeUtils_1.getDefaultString(messageKey, localizeUtils_1.getDefaultString(additionalMessageKey), envList, filePath),
            displayMessage: localizeUtils_1.getLocalizedString(messageKey, localizeUtils_1.getLocalizedString(additionalMessageKey), envList, filePath),
            helpLink: helpLink,
        });
    }
}
exports.MissingEnvInFileUserError = MissingEnvInFileUserError;
