// @microsoft/teamsfx-core/build/component/feature/bot/constants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.NotificationTriggers = exports.QuestionNames = exports.TriggerTemplateScenarioMappings = exports.TemplateProjectsScenarios = exports.BotCapabilities = void 0;
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT license.
const constants_1 = require("../../constants");
exports.BotCapabilities = {
    NOTIFICATION: "notification",
    COMMAND_AND_RESPONSE: "command-response",
    WORKFLOW: "workflow",
    BOT: "bot",
    MESSAGE_EXTENSION: "message-extension",
    M365_SEARCH_APP: "m365-search-app",
};
var TemplateProjectsScenarios;
(function (TemplateProjectsScenarios) {
    TemplateProjectsScenarios["DEFAULT_SCENARIO_NAME"] = "default";
    TemplateProjectsScenarios["NOTIFICATION_RESTIFY_SCENARIO_NAME"] = "notification-restify";
    TemplateProjectsScenarios["NOTIFICATION_WEBAPI_SCENARIO_NAME"] = "notification-webapi";
    TemplateProjectsScenarios["NOTIFICATION_FUNCTION_BASE_SCENARIO_NAME"] = "notification-function-base";
    TemplateProjectsScenarios["NOTIFICATION_FUNCTION_TRIGGER_HTTP_SCENARIO_NAME"] = "notification-trigger-http";
    TemplateProjectsScenarios["NOTIFICATION_FUNCTION_TRIGGER_TIMER_SCENARIO_NAME"] = "notification-trigger-timer";
    TemplateProjectsScenarios["COMMAND_AND_RESPONSE_SCENARIO_NAME"] = "command-and-response";
    TemplateProjectsScenarios["WORKFLOW_SCENARIO_NAME"] = "workflow";
    TemplateProjectsScenarios["M365_SCENARIO_NAME"] = "m365";
})(TemplateProjectsScenarios = exports.TemplateProjectsScenarios || (exports.TemplateProjectsScenarios = {}));
exports.TriggerTemplateScenarioMappings = {
    [constants_1.BotNotificationTriggers.Http]: TemplateProjectsScenarios.NOTIFICATION_FUNCTION_TRIGGER_HTTP_SCENARIO_NAME,
    [constants_1.BotNotificationTriggers.Timer]: TemplateProjectsScenarios.NOTIFICATION_FUNCTION_TRIGGER_TIMER_SCENARIO_NAME,
};
class QuestionNames {
}
exports.QuestionNames = QuestionNames;
QuestionNames.CAPABILITIES = "capabilities";
QuestionNames.BOT_HOST_TYPE_TRIGGER = "bot-host-type-trigger";
exports.NotificationTriggers = {
    HTTP: "http",
    TIMER: "timer",
};
