// @microsoft/teamsfx-core/build/component/generator/constant.js
"use strict";
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT license.
Object.defineProperty(exports, "__esModule", { value: true });
exports.placeholderDelimiters = exports.commonTemplateName = exports.componentName = exports.errorSource = exports.templateAlphaVersion = exports.templateFileExt = exports.sampleDefaultTimeoutInMs = exports.defaultTimeoutInMs = exports.defaultTryLimits = exports.templateDownloadBaseUrl = exports.templateTagListUrl = void 0;
exports.templateTagListUrl = "https://github.com/OfficeDev/TeamsFx/releases/download/template-tag-list/template-tags.txt";
exports.templateDownloadBaseUrl = "https://github.com/OfficeDev/TeamsFx/releases/download";
exports.defaultTryLimits = 1;
exports.defaultTimeoutInMs = 20000;
exports.sampleDefaultTimeoutInMs = 40000;
exports.templateFileExt = ".tpl";
exports.templateAlphaVersion = "templates-0.0.0-alpha";
exports.errorSource = "GEN";
exports.componentName = "generator";
exports.commonTemplateName = "common";
exports.placeholderDelimiters = ["{%", "%}"];
