// @microsoft/teamsfx-core/build/component/generator/spfxGenerator.js
"use strict";
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT license.
Object.defineProperty(exports, "__esModule", { value: true });
exports.SPFxGenerator = void 0;
const tslib_1 = require("tslib");
const lib_1 = require("@feathersjs/hooks/lib");
const teamsfx_api_1 = require("@microsoft/teamsfx-api");
const path = tslib_1.__importStar(require("path"));
const fs_extra_1 = tslib_1.__importDefault(require("fs-extra"));
const actionExecutionMW_1 = require("../middleware/actionExecutionMW");
const progress_helper_1 = require("../resource/spfx/utils/progress-helper");
const questions_1 = require("../resource/spfx/utils/questions");
const error_1 = require("../resource/spfx/error");
const utils_1 = require("../resource/spfx/utils/utils");
const lodash_1 = require("lodash");
const constants_1 = require("../resource/spfx/utils/constants");
const yoChecker_1 = require("../resource/spfx/depsChecker/yoChecker");
const generatorChecker_1 = require("../resource/spfx/depsChecker/generatorChecker");
const tools_1 = require("../../common/tools");
const deps_checker_1 = require("../../common/deps-checker");
const telemetryEvents_1 = require("../resource/spfx/utils/telemetryEvents");
const generator_1 = require("./generator");
const question_1 = require("../../core/question");
const localizeUtils_1 = require("../../common/localizeUtils");
class SPFxGenerator {
    static async generate(context, inputs, destinationPath) {
        const yeomanRes = await this.doYeomanScaffold(context, inputs, destinationPath);
        if (yeomanRes.isErr())
            return teamsfx_api_1.err(yeomanRes.error);
        const templateRes = await generator_1.Generator.generateTemplate(context, destinationPath, constants_1.Constants.TEMPLATE_NAME, "ts");
        if (templateRes.isErr())
            return teamsfx_api_1.err(templateRes.error);
        return teamsfx_api_1.ok(undefined);
    }
    static async doYeomanScaffold(context, inputs, destinationPath) {
        var _a, _b, _c, _d, _e;
        const ui = context.userInteraction;
        const progressHandler = await progress_helper_1.ProgressHelper.startScaffoldProgressHandler(ui);
        try {
            const webpartName = inputs[questions_1.SPFXQuestionNames.webpart_name];
            const framework = inputs[questions_1.SPFXQuestionNames.framework_type];
            const solutionName = inputs[question_1.CoreQuestionNames.AppName];
            const componentName = utils_1.Utils.normalizeComponentName(webpartName);
            const componentNameCamelCase = lodash_1.camelCase(componentName);
            await (progressHandler === null || progressHandler === void 0 ? void 0 : progressHandler.next(localizeUtils_1.getLocalizedString("plugins.spfx.scaffold.dependencyCheck")));
            const yoChecker = new yoChecker_1.YoChecker(context.logProvider);
            const spGeneratorChecker = new generatorChecker_1.GeneratorChecker(context.logProvider);
            const yoInstalled = await yoChecker.isInstalled();
            const generatorInstalled = await spGeneratorChecker.isInstalled();
            if (!yoInstalled || !generatorInstalled) {
                await (progressHandler === null || progressHandler === void 0 ? void 0 : progressHandler.next(localizeUtils_1.getLocalizedString("plugins.spfx.scaffold.dependencyInstall")));
                if (tools_1.isYoCheckerEnabled()) {
                    const yoRes = await yoChecker.ensureDependency(context);
                    if (yoRes.isErr()) {
                        throw error_1.DependencyInstallError("yo");
                    }
                }
                if (tools_1.isGeneratorCheckerEnabled()) {
                    const spGeneratorRes = await spGeneratorChecker.ensureDependency(context);
                    if (spGeneratorRes.isErr()) {
                        throw error_1.DependencyInstallError("sharepoint generator");
                    }
                }
            }
            await (progressHandler === null || progressHandler === void 0 ? void 0 : progressHandler.next(localizeUtils_1.getLocalizedString("plugins.spfx.scaffold.scaffoldProject")));
            if (inputs.platform === teamsfx_api_1.Platform.VSCode) {
                context.logProvider.outputChannel.show();
            }
            const yoEnv = process.env;
            if (yoEnv.PATH) {
                yoEnv.PATH = tools_1.isYoCheckerEnabled()
                    ? `${await (await yoChecker.getBinFolders()).join(path.delimiter)}${path.delimiter}${(_a = process.env.PATH) !== null && _a !== void 0 ? _a : ""}`
                    : process.env.PATH;
            }
            else {
                yoEnv.Path = tools_1.isYoCheckerEnabled()
                    ? `${await (await yoChecker.getBinFolders()).join(path.delimiter)}${path.delimiter}${(_b = process.env.Path) !== null && _b !== void 0 ? _b : ""}`
                    : process.env.Path;
            }
            const args = [
                tools_1.isGeneratorCheckerEnabled()
                    ? spGeneratorChecker.getSpGeneratorPath()
                    : "@microsoft/sharepoint",
                "--skip-install",
                "true",
                "--component-type",
                "webpart",
                "--component-name",
                webpartName,
                "--environment",
                "spo",
                "--skip-feature-deployment",
                "true",
                "--is-domain-isolated",
                "false",
            ];
            if (framework) {
                args.push("--framework", framework);
            }
            if (solutionName) {
                args.push("--solution-name", solutionName);
            }
            await deps_checker_1.cpUtils.executeCommand(destinationPath, context.logProvider, {
                timeout: 2 * 60 * 1000,
                env: yoEnv,
            }, "yo", ...args);
            const newPath = path.join(destinationPath, "src");
            const currentPath = path.join(destinationPath, solutionName);
            await fs_extra_1.default.rename(currentPath, newPath);
            await (progressHandler === null || progressHandler === void 0 ? void 0 : progressHandler.next(localizeUtils_1.getLocalizedString("plugins.spfx.scaffold.updateManifest")));
            const manifestPath = `${newPath}/src/webparts/${componentNameCamelCase}/${componentName}WebPart.manifest.json`;
            const manifest = await fs_extra_1.default.readFile(manifestPath, "utf8");
            let manifestString = manifest.toString();
            manifestString = manifestString.replace(`"supportedHosts": ["SharePointWebPart"]`, `"supportedHosts": ["SharePointWebPart", "TeamsPersonalApp", "TeamsTab"]`);
            await fs_extra_1.default.writeFile(manifestPath, manifestString);
            const matchHashComment = new RegExp(/(\/\/ .*)/, "gi");
            const manifestJson = JSON.parse(manifestString.replace(matchHashComment, "").trim());
            const componentId = manifestJson.id;
            if (!context.templateVariables) {
                context.templateVariables = generator_1.Generator.getDefaultVariables(solutionName);
            }
            context.templateVariables["componentId"] = componentId;
            context.templateVariables["webpartName"] = webpartName;
            // remove dataVersion() function, related issue: https://github.com/SharePoint/sp-dev-docs/issues/6469
            const webpartFile = `${newPath}/src/webparts/${componentNameCamelCase}/${componentName}WebPart.ts`;
            const codeFile = await fs_extra_1.default.readFile(webpartFile, "utf8");
            let codeString = codeFile.toString();
            codeString = codeString.replace(`  protected get dataVersion(): Version {\r\n    return Version.parse('1.0');\r\n  }\r\n\r\n`, ``);
            codeString = codeString.replace(`import { Version } from '@microsoft/sp-core-library';\r\n`, ``);
            await fs_extra_1.default.writeFile(webpartFile, codeString);
            // remove .vscode
            const debugPath = `${newPath}/.vscode`;
            if (await fs_extra_1.default.pathExists(debugPath)) {
                await fs_extra_1.default.remove(debugPath);
            }
            await (progressHandler === null || progressHandler === void 0 ? void 0 : progressHandler.end(true));
            return teamsfx_api_1.ok(undefined);
        }
        catch (error) {
            if (error.name === "DependencyInstallFailed") {
                const globalYoVersion = utils_1.Utils.getPackageVersion("yo");
                const globalGenVersion = utils_1.Utils.getPackageVersion("@microsoft/generator-sharepoint");
                const yoInfo = yoChecker_1.YoChecker.getDependencyInfo();
                const genInfo = generatorChecker_1.GeneratorChecker.getDependencyInfo();
                const yoMessage = globalYoVersion === undefined
                    ? "    yo not installed"
                    : `    globally installed yo@${globalYoVersion}`;
                const generatorMessage = globalGenVersion === undefined
                    ? "    @microsoft/generator-sharepoint not installed"
                    : `    globally installed @microsoft/generator-sharepoint@${globalYoVersion}`;
                (_c = context.logProvider) === null || _c === void 0 ? void 0 : _c.error(`We've encountered some issues when trying to install prerequisites under HOME/.fx folder.  Learn how to remediate by going to this link(aka.ms/teamsfx-spfx-help) and following the steps applicable to your system: \n ${yoMessage} \n ${generatorMessage}`);
                (_d = context.logProvider) === null || _d === void 0 ? void 0 : _d.error(`Teams Toolkit recommends using ${yoInfo.displayName} ${genInfo.displayName}`);
            }
            if (error.message &&
                error.message.includes("'yo' is not recognized as an internal or external command")) {
                (_e = context.logProvider) === null || _e === void 0 ? void 0 : _e.error("NPM v6.x with Node.js v12.13.0+ (Erbium) or Node.js v14.15.0+ (Fermium) is recommended for spfx scaffolding and later development. You can use correct version and try again.");
            }
            await (progressHandler === null || progressHandler === void 0 ? void 0 : progressHandler.end(false));
            return teamsfx_api_1.err(error_1.ScaffoldError(error));
        }
    }
}
tslib_1.__decorate([
    lib_1.hooks([
        actionExecutionMW_1.ActionExecutionMW({
            enableTelemetry: true,
            telemetryComponentName: constants_1.Constants.PLUGIN_DEV_NAME,
            telemetryEventName: telemetryEvents_1.TelemetryEvents.Generate,
            errorSource: constants_1.Constants.PLUGIN_DEV_NAME,
        }),
    ]),
    tslib_1.__metadata("design:type", Function),
    tslib_1.__metadata("design:paramtypes", [Object, Object, String]),
    tslib_1.__metadata("design:returntype", Promise)
], SPFxGenerator, "generate", null);
exports.SPFxGenerator = SPFxGenerator;
