// @microsoft/teamsfx-core/build/component/driver/m365/error/invalidParameterUserError.js
"use strict";
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT license.
Object.defineProperty(exports, "__esModule", { value: true });
exports.InvalidParameterUserError = void 0;
const teamsfx_api_1 = require("@microsoft/teamsfx-api");
const localizeUtils_1 = require("../../../../common/localizeUtils");
const errorCode = "InvalidParameter";
const messageKey = "driver.m365.error.invalidParameter";
class InvalidParameterUserError extends teamsfx_api_1.UserError {
    constructor(actionName, invalidParameters, helpLink) {
        let parameterList;
        if (Array.isArray(invalidParameters)) {
            parameterList = invalidParameters.join(", ");
        }
        else {
            parameterList = invalidParameters;
        }
        super({
            source: actionName,
            name: errorCode,
            message: localizeUtils_1.getDefaultString(messageKey, actionName, parameterList),
            displayMessage: localizeUtils_1.getLocalizedString(messageKey, actionName, parameterList),
            helpLink: helpLink,
        });
    }
}
exports.InvalidParameterUserError = InvalidParameterUserError;
