// @microsoft/teamsfx-core/build/common/azure-hosting/azureAppServiceHosting.js
"use strict";
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT license.
Object.defineProperty(exports, "__esModule", { value: true });
exports.AzureAppServiceHosting = void 0;
const azureHosting_1 = require("./azureHosting");
const interfaces_1 = require("./interfaces");
const teamsfx_api_1 = require("@microsoft/teamsfx-api");
const utils_1 = require("./utils");
const resourceId = "provisionOutputs.webAppOutput.value.webAppResourceId";
const hostName = "provisionOutputs.webAppOutput.value.validDomain";
const webAppEndpoint = "provisionOutputs.webAppOutput.value.siteEndpoint";
const endpointAsParam = "webAppProvision.outputs.webAppEndpoint";
class AzureAppServiceHosting extends azureHosting_1.AzureHosting {
    constructor() {
        super(...arguments);
        this.configurable = true;
        this.hostType = interfaces_1.ServiceType.AppService;
        this.reference = {
            resourceId: resourceId,
            hostName: hostName,
            webAppEndpoint: webAppEndpoint,
            endpointAsParam: endpointAsParam,
        };
    }
    async deploy(resourceId, tokenProvider, buffer) {
        await super.deploy(resourceId, tokenProvider, buffer);
        await utils_1.azureWebSiteDeploy(resourceId, tokenProvider, buffer, this.logger);
        return teamsfx_api_1.Void;
    }
}
exports.AzureAppServiceHosting = AzureAppServiceHosting;
