// @microsoft/teamsfx-core/build/common/azure-hosting/hostingConstant.js
"use strict";
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT license.
Object.defineProperty(exports, "__esModule", { value: true });
exports.DeployConfigsConstants = exports.AzureOperationCommonConstants = exports.ErrorNameConstant = exports.AzureOpsConstant = exports.DeployStatusConstant = exports.AppServiceBicepConstant = void 0;
const localizeUtils_1 = require("../localizeUtils");
class AppServiceBicepConstant {
}
exports.AppServiceBicepConstant = AppServiceBicepConstant;
AppServiceBicepConstant.resourceId = "provisionOutputs.botOutput.value.resourceId";
AppServiceBicepConstant.hostName = "provisionOutputs.botOutput.value.validDomain";
AppServiceBicepConstant.webAppEndpoint = "provisionOutputs.botOutputs.value.botWebAppEndpoint";
class DeployStatusConstant {
}
exports.DeployStatusConstant = DeployStatusConstant;
DeployStatusConstant.RETRY_TIMES = 120; // Timeout: 20 min
DeployStatusConstant.BACKOFF_TIME_S = 10;
class AzureOpsConstant {
}
exports.AzureOpsConstant = AzureOpsConstant;
AzureOpsConstant.TRY_LOGIN_AZURE = () => localizeUtils_1.getLocalizedString("plugin.hosting.LoginToAzure");
AzureOpsConstant.CHECK_OUTPUT_LOG_AND_TRY_TO_FIX = () => localizeUtils_1.getLocalizedString("plugin.hosting.CheckLogAndFix");
AzureOpsConstant.RETRY_CURRENT_STEP = () => localizeUtils_1.getLocalizedString("suggestions.retryTheCurrentStep");
AzureOpsConstant.FAIL_TO_GET_AZURE_CREDENTIALS = () => [
    localizeUtils_1.getDefaultString("plugin.hosting.FailRetrieveAzureCredentials"),
    localizeUtils_1.getLocalizedString("plugin.hosting.FailRetrieveAzureCredentials"),
];
AzureOpsConstant.FAIL_TO_UPDATE_MESSAGE_ENDPOINT = (endpoint) => [
    localizeUtils_1.getDefaultString("plugin.hosting.FailedUpdateMessageEndpoint", endpoint),
    localizeUtils_1.getLocalizedString("plugin.hosting.FailedUpdateMessageEndpoint", endpoint),
];
AzureOpsConstant.FAIL_TO_PROVISION_SOME_RESOURCE = (resource) => [
    localizeUtils_1.getDefaultString("plugin.hosting.FailedToProvision", resource),
    localizeUtils_1.getLocalizedString("plugin.hosting.FailedToProvision", resource),
];
AzureOpsConstant.FAIL_TO_LIST_PUBLISHING_CREDENTIALS = () => [
    localizeUtils_1.getDefaultString("plugin.hosting.FailedListPublishingCredentials"),
    localizeUtils_1.getLocalizedString("plugin.hosting.FailedListPublishingCredentials"),
];
AzureOpsConstant.FAIL_TO_DO_ZIP_DEPLOY = () => [
    localizeUtils_1.getDefaultString("plugin.hosting.FailedDeployZipFile"),
    localizeUtils_1.getLocalizedString("plugin.hosting.FailedDeployZipFile"),
];
AzureOpsConstant.FAIL_TO_CHECK_DEPLOY_STATUS = () => [
    // eslint-disable-next-line no-secrets/no-secrets
    localizeUtils_1.getDefaultString("plugin.hosting.FailedCheckDeployStatus"),
    // eslint-disable-next-line no-secrets/no-secrets
    localizeUtils_1.getLocalizedString("plugin.hosting.FailedCheckDeployStatus"),
];
AzureOpsConstant.CHECK_DEPLOY_STATUS_TIMEOUT = () => [
    // eslint-disable-next-line no-secrets/no-secrets
    localizeUtils_1.getDefaultString("plugin.hosting.CheckDeployStatusTimeout"),
    // eslint-disable-next-line no-secrets/no-secrets
    localizeUtils_1.getLocalizedString("plugin.hosting.CheckDeployStatusTimeout"),
];
AzureOpsConstant.FAIL_TO_RESTART_APP_SERVICE = () => [
    localizeUtils_1.getDefaultString("plugin.hosting.FailedRestartWebApp"),
    localizeUtils_1.getLocalizedString("plugin.hosting.FailedRestartWebApp"),
];
class ErrorNameConstant {
}
exports.ErrorNameConstant = ErrorNameConstant;
ErrorNameConstant.PRECONDITION_ERROR = "PreconditionError";
ErrorNameConstant.MSG_ENDPOINT_UPDATING_ERROR = "MessageEndpointUpdatingError";
ErrorNameConstant.PROVISION_ERROR = "ProvisionError";
ErrorNameConstant.LIST_PUBLISHING_CREDENTIALS_ERROR = "ListPublishingCredentialsError";
ErrorNameConstant.ZIP_DEPLOY_ERROR = "ZipDeployError";
ErrorNameConstant.DEPLOY_STATUS_ERROR = "DeployStatusError";
ErrorNameConstant.DEPLOY_TIMEOUT_ERROR = "DeployTimeoutError";
ErrorNameConstant.RESTART_WEBAPP_ERROR = "RestartWebappError";
class AzureOperationCommonConstants {
}
exports.AzureOperationCommonConstants = AzureOperationCommonConstants;
AzureOperationCommonConstants.msInOneSecond = 1000;
AzureOperationCommonConstants.zipTimeMSGranularity = 2 * AzureOperationCommonConstants.msInOneSecond;
AzureOperationCommonConstants.latestTrustMtime = new Date(2000, 1, 1).getTime();
AzureOperationCommonConstants.deployTimeoutInMs = 10 * 60 * 1000;
class DeployConfigsConstants {
}
exports.DeployConfigsConstants = DeployConfigsConstants;
DeployConfigsConstants.DEPLOYMENT_FOLDER = ".deployment";
DeployConfigsConstants.DEPLOYMENT_INFO_FILE = "deployment.json";
DeployConfigsConstants.DEPLOYMENT_ZIP_CACHE_FILE = "deployment.zip";
DeployConfigsConstants.FUNC_IGNORE_FILE = ".funcignore";
DeployConfigsConstants.GIT_IGNORE_FILE = ".gitignore";
