// @microsoft/teamsfx-core/build/common/azure-hosting/hostingError.js
"use strict";
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT license.
Object.defineProperty(exports, "__esModule", { value: true });
exports.RestartWebAppError = exports.DeployTimeoutError = exports.DeployStatusError = exports.ZipDeployError = exports.ListPublishingCredentialsError = exports.ProvisionError = exports.MessageEndpointUpdatingError = exports.PreconditionError = exports.CommonHostingError = void 0;
const localizeUtils_1 = require("../localizeUtils");
const hostingConstant_1 = require("./hostingConstant");
class CommonHostingError extends Error {
    constructor(name, details, innerError, suggestions) {
        super(details[0]);
        this.name = name;
        this.details = details;
        this.suggestions = suggestions !== null && suggestions !== void 0 ? suggestions : [
            hostingConstant_1.AzureOpsConstant.CHECK_OUTPUT_LOG_AND_TRY_TO_FIX(),
            hostingConstant_1.AzureOpsConstant.RETRY_CURRENT_STEP(),
        ];
        this.innerError = innerError;
    }
    genMessage() {
        let msg = `${this.details[0]} `;
        if (this.suggestions.length > 0) {
            msg += localizeUtils_1.getDefaultString("plugins.common.hosting.ErrorSuggestions", this.suggestions.join(" "));
        }
        return msg;
    }
    genDisplayMessage() {
        let msg = `${this.details[1]} `;
        if (this.suggestions.length > 0) {
            msg += localizeUtils_1.getLocalizedString("plugins.common.hosting.ErrorSuggestions", this.suggestions.join(" "));
        }
        return msg;
    }
}
exports.CommonHostingError = CommonHostingError;
class PreconditionError extends CommonHostingError {
    constructor(message, suggestions) {
        super(hostingConstant_1.ErrorNameConstant.PRECONDITION_ERROR, message, suggestions);
    }
}
exports.PreconditionError = PreconditionError;
class MessageEndpointUpdatingError extends CommonHostingError {
    constructor(endpoint, innerError) {
        super(hostingConstant_1.ErrorNameConstant.MSG_ENDPOINT_UPDATING_ERROR, hostingConstant_1.AzureOpsConstant.FAIL_TO_UPDATE_MESSAGE_ENDPOINT(endpoint), innerError);
    }
}
exports.MessageEndpointUpdatingError = MessageEndpointUpdatingError;
class ProvisionError extends CommonHostingError {
    constructor(resource, innerError) {
        super(hostingConstant_1.ErrorNameConstant.PROVISION_ERROR, hostingConstant_1.AzureOpsConstant.FAIL_TO_PROVISION_SOME_RESOURCE(resource), innerError);
    }
}
exports.ProvisionError = ProvisionError;
class ListPublishingCredentialsError extends CommonHostingError {
    constructor(innerError) {
        super(hostingConstant_1.ErrorNameConstant.LIST_PUBLISHING_CREDENTIALS_ERROR, hostingConstant_1.AzureOpsConstant.FAIL_TO_LIST_PUBLISHING_CREDENTIALS(), innerError);
    }
}
exports.ListPublishingCredentialsError = ListPublishingCredentialsError;
class ZipDeployError extends CommonHostingError {
    constructor(innerError) {
        super(hostingConstant_1.ErrorNameConstant.ZIP_DEPLOY_ERROR, hostingConstant_1.AzureOpsConstant.FAIL_TO_DO_ZIP_DEPLOY(), innerError);
    }
}
exports.ZipDeployError = ZipDeployError;
class DeployStatusError extends CommonHostingError {
    constructor(innerError) {
        super(hostingConstant_1.ErrorNameConstant.DEPLOY_STATUS_ERROR, hostingConstant_1.AzureOpsConstant.FAIL_TO_CHECK_DEPLOY_STATUS(), innerError);
    }
}
exports.DeployStatusError = DeployStatusError;
class DeployTimeoutError extends CommonHostingError {
    constructor() {
        super(hostingConstant_1.ErrorNameConstant.DEPLOY_TIMEOUT_ERROR, hostingConstant_1.AzureOpsConstant.CHECK_DEPLOY_STATUS_TIMEOUT());
    }
}
exports.DeployTimeoutError = DeployTimeoutError;
class RestartWebAppError extends CommonHostingError {
    constructor(innerError) {
        super(hostingConstant_1.ErrorNameConstant.RESTART_WEBAPP_ERROR, hostingConstant_1.AzureOpsConstant.FAIL_TO_RESTART_APP_SERVICE(), innerError);
    }
}
exports.RestartWebAppError = RestartWebAppError;
