// @microsoft/teamsfx-core/build/component/resource/azureSql/utils/common.js
"use strict";
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT license.
Object.defineProperty(exports, "__esModule", { value: true });
exports.formatEndpoint = exports.parseToken = exports.UserType = void 0;
const tslib_1 = require("tslib");
const jwt_decode_1 = tslib_1.__importDefault(require("jwt-decode"));
const constants_1 = require("../constants");
var UserType;
(function (UserType) {
    UserType["User"] = "User";
    UserType["ServicePrincipal"] = "ServicePrincipal";
})(UserType = exports.UserType || (exports.UserType = {}));
function parseToken(accessToken) {
    const jwt = jwt_decode_1.default(accessToken);
    let authType;
    if (jwt.ver === constants_1.Constants.jwtToken.ver1) {
        authType = jwt.appidacr;
    }
    else if (jwt.ver === constants_1.Constants.jwtToken.ver2) {
        authType = jwt.azpacr;
    }
    else {
        throw new Error("invalid token");
    }
    if (authType === constants_1.Constants.jwtToken.userType) {
        return {
            name: jwt.name,
            objectId: jwt.oid,
            userType: UserType.User,
        };
    }
    else {
        return {
            name: jwt.appid,
            objectId: jwt.oid,
            userType: UserType.ServicePrincipal,
        };
    }
}
exports.parseToken = parseToken;
function formatEndpoint(endpoint) {
    endpoint = endpoint.toLowerCase();
    endpoint = endpoint.replace(/[^a-z0-9-]/gi, "");
    if (endpoint[0] === "-") {
        endpoint = endpoint.slice(1);
    }
    return endpoint;
}
exports.formatEndpoint = formatEndpoint;
