// @microsoft/teamsfx-core/build/component/resource/simpleAuth/result.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ResultFactory = void 0;
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT license.
const teamsfx_api_1 = require("@microsoft/teamsfx-api");
const constants_1 = require("./constants");
class ResultFactory {
    static UserError(name, messages, innerError, stack, helpLink) {
        return new teamsfx_api_1.UserError({
            name,
            message: messages[0],
            displayMessage: messages[1],
            source: this.source,
            helpLink,
            error: innerError,
        });
    }
    static SystemError(name, messages, innerError, stack, issueLink) {
        return new teamsfx_api_1.SystemError({
            name,
            message: messages[0],
            displayMessage: messages[1],
            source: this.source,
            issueLink,
            error: innerError,
        });
    }
    static Success(result) {
        return teamsfx_api_1.ok(result);
    }
}
exports.ResultFactory = ResultFactory;
ResultFactory.source = constants_1.Constants.SimpleAuthPlugin.shortName;
