// @microsoft/teamsfx-core/build/component/error/componentError.js
"use strict";
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT license.
Object.defineProperty(exports, "__esModule", { value: true });
exports.PrerequisiteError = exports.ExecuteCommandError = exports.ExternalApiCallError = exports.BaseComponentInnerError = void 0;
const teamsfx_api_1 = require("@microsoft/teamsfx-api");
const localizeUtils_1 = require("../../common/localizeUtils");
const deployConstant_1 = require("../constant/deployConstant");
/**
 * component error
 * store more information in FxError's innerError property for telemetry, logging and debugging
 *
 * @property name name of error, for classification
 * @property message always save error message in english
 * @property errorType which type of error will throw to fx-core
 * @property displayMessage the message show to user, always save as localized message
 * @property helpLink link to help page, mostly no use
 * @property detail detail error message, http response, console output, etc..
 */
class BaseComponentInnerError extends Error {
    constructor(source, errorType, name, messageKey, messageParams, suggestionKey, detail, helpLink) {
        super(messageKey
            ? messageParams
                ? localizeUtils_1.getDefaultString(messageKey, ...messageParams)
                : localizeUtils_1.getDefaultString(messageKey)
            : "");
        this.source = source;
        this.errorType = errorType;
        this.name = name;
        this.helpLink = helpLink;
        this.suggestionKey = suggestionKey;
        this.displayMessage = messageKey
            ? messageParams
                ? localizeUtils_1.getLocalizedString(messageKey, ...messageParams)
                : localizeUtils_1.getLocalizedString(messageKey)
            : "";
        this.detail = detail;
    }
    toFxError() {
        if (this.errorType === "UserError") {
            return new teamsfx_api_1.UserError({
                source: this.source,
                error: this,
                helpLink: this.helpLink,
                name: this.name,
                message: this.message,
                displayMessage: this.toDisplayMessage(),
            });
        }
        else {
            return new teamsfx_api_1.SystemError({
                source: this.source,
                name: this.name,
                message: this.message,
                error: this,
                helpLink: this.helpLink,
                displayMessage: this.toDisplayMessage(),
            });
        }
    }
    /**
     * The suggestion message will be localized and add 'Suggestions:' by default
     * the final method will be like:
     *    'Error message. Suggestions: suggestion message'
     * can be overwritten by subclasses
     * @protected
     */
    toDisplayMessage() {
        var _a;
        const suggestion = (_a = this.suggestionKey) === null || _a === void 0 ? void 0 : _a.map((key) => localizeUtils_1.getLocalizedString(key)).join(" ");
        return this.displayMessage
            ? suggestion
                ? this.displayMessage + localizeUtils_1.getLocalizedString("plugins.bot.ErrorSuggestions", suggestion)
                : this.displayMessage
            : this.message;
    }
    static unknownError(source, error) {
        return new BaseComponentInnerError(source, "SystemError", "UnknownError", "driver.deploy.error.unknownError", [JSON.stringify(error)]);
    }
}
exports.BaseComponentInnerError = BaseComponentInnerError;
/**
 * component call external api error
 */
class ExternalApiCallError extends BaseComponentInnerError {
    constructor(source, name, message, statusCode, messageParams, suggestionKey, detail, helpLink) {
        super(source, "UserError", name, message, messageParams, suggestionKey, detail, helpLink);
        this.statusCode = statusCode;
    }
    static getAzureCredentialError(source, error) {
        error = error !== null && error !== void 0 ? error : "";
        return new ExternalApiCallError(source, "GetAzureCredentialError", "plugins.bot.FailRetrieveAzureCredentials", -1, undefined, undefined, typeof error === "string" ? error : JSON.stringify(error));
    }
    static getSasTokenError(source, detail) {
        return new ExternalApiCallError(source, "AzureStorageSASToeknEmpty", "error.frontend.GetContainerError", -1, [deployConstant_1.DeployConstant.AZURE_STORAGE_CONTAINER_NAME], [
            "plugins.frontend.checkSystemTimeTip",
            // eslint-disable-next-line no-secrets/no-secrets
            "plugins.frontend.checkStoragePermissionsTip",
            "plugins.frontend.checkNetworkTip",
        ], detail);
    }
}
exports.ExternalApiCallError = ExternalApiCallError;
class ExecuteCommandError extends BaseComponentInnerError {
    constructor(source, name, messageKey, messageParams, error) {
        super(source, "UserError", name, messageKey, messageParams, ["plugins.bot.CheckCommandOutput", "suggestions.retryTheCurrentStep"], error);
    }
    static fromErrorOutput(source, commands, error) {
        return new ExecuteCommandError(source, "CommandExecutionError", "plugins.bot.RunFailedCommand", commands, typeof error === "string" ? error : JSON.stringify(error));
    }
}
exports.ExecuteCommandError = ExecuteCommandError;
/**
 * parameter in environmental error or user input error
 */
class PrerequisiteError extends BaseComponentInnerError {
    constructor(source, name, messageKey, messageParams, suggestionKey, helpLink) {
        super(source, "UserError", name, messageKey, messageParams, suggestionKey, undefined, helpLink);
    }
    static somethingIllegal(source, name, messageKey, messageParams, helpLink) {
        return new PrerequisiteError(source, "Illegal" + PrerequisiteError.toCamel(name), messageKey, messageParams, undefined, helpLink);
    }
    static somethingMissing(source, name, helpLink) {
        return new PrerequisiteError(source, PrerequisiteError.toCamel(name) + "IsMissing", "plugins.bot.SomethingIsMissing", [name], undefined, helpLink);
    }
    static folderNotExists(source, name, helpLink) {
        return new PrerequisiteError(source, "FolderNotExists", "driver.env.error.folderNotExist", [name], undefined, helpLink);
    }
    static toCamel(name) {
        return name.charAt(0).toUpperCase() + name.slice(1);
    }
}
exports.PrerequisiteError = PrerequisiteError;
