// @microsoft/teamsfx-core/build/component/debug/util/launchTransparency.js
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT license.
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.generateSpfxCompounds = exports.generateSpfxConfigurations = exports.generateM365Compounds = exports.generateM365Configurations = exports.generateCompounds = exports.generateConfigurations = exports.generateM365LaunchJson = exports.generateLaunchJson = void 0;
const tslib_1 = require("tslib");
const constants_1 = require("../../../common/local/constants");
const Launch = tslib_1.__importStar(require("./launch"));
const LaunchNext = tslib_1.__importStar(require("./launchNext"));
function generateLaunchJson(includeFrontend, includeBackend, includeBot) {
    return {
        version: "0.2.0",
        configurations: generateConfigurations(includeFrontend, includeBackend, includeBot),
        compounds: generateCompounds(includeFrontend, includeBackend, includeBot),
    };
}
exports.generateLaunchJson = generateLaunchJson;
function generateM365LaunchJson(includeFrontend, includeBackend, includeBot) {
    return {
        version: "0.2.0",
        configurations: generateM365Configurations(includeFrontend, includeBackend, includeBot),
        compounds: generateM365Compounds(includeFrontend, includeBackend, includeBot),
    };
}
exports.generateM365LaunchJson = generateM365LaunchJson;
function generateConfigurations(includeFrontend, includeBackend, includeBot) {
    return LaunchNext.generateConfigurations(includeFrontend, includeBackend, includeBot);
}
exports.generateConfigurations = generateConfigurations;
function generateCompounds(includeFrontend, includeBackend, includeBot) {
    const result = LaunchNext.generateCompounds(includeFrontend, includeBackend, includeBot);
    result[0].preLaunchTask = constants_1.TaskOverallLabel.TransparentDefault;
    result[1].preLaunchTask = constants_1.TaskOverallLabel.TransparentDefault;
    return result;
}
exports.generateCompounds = generateCompounds;
function generateM365Configurations(includeFrontend, includeBackend, includeBot) {
    return LaunchNext.generateM365Configurations(includeFrontend, includeBackend, includeBot);
}
exports.generateM365Configurations = generateM365Configurations;
function generateM365Compounds(includeFrontend, includeBackend, includeBot) {
    const result = LaunchNext.generateM365Compounds(includeFrontend, includeBackend, includeBot);
    result[0].preLaunchTask = constants_1.TaskOverallLabel.TransparentDefault;
    result[1].preLaunchTask = constants_1.TaskOverallLabel.TransparentDefault;
    result[2].preLaunchTask = constants_1.TaskOverallLabel.TransparentM365;
    result[3].preLaunchTask = constants_1.TaskOverallLabel.TransparentM365;
    if (includeFrontend) {
        result[4].preLaunchTask = constants_1.TaskOverallLabel.TransparentM365;
        result[5].preLaunchTask = constants_1.TaskOverallLabel.TransparentM365;
    }
    return result;
}
exports.generateM365Compounds = generateM365Compounds;
function generateSpfxConfigurations() {
    return Launch.generateSpfxConfigurations();
}
exports.generateSpfxConfigurations = generateSpfxConfigurations;
function generateSpfxCompounds() {
    const result = Launch.generateSpfxCompounds();
    result[0].preLaunchTask = constants_1.TaskOverallLabel.TransparentDefault;
    result[1].preLaunchTask = constants_1.TaskOverallLabel.TransparentDefault;
    return result;
}
exports.generateSpfxCompounds = generateSpfxCompounds;
