// @microsoft/teamsfx-core/build/component/driver/botAadApp/error/unhandledError.js
"use strict";
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT license.
Object.defineProperty(exports, "__esModule", { value: true });
exports.UnhandledSystemError = exports.UnhandledUserError = void 0;
const teamsfx_api_1 = require("@microsoft/teamsfx-api");
const localizeUtils_1 = require("../../../../common/localizeUtils");
const errorCode = "UnhandledError";
const messageKey = "driver.aadApp.error.unhandledError";
class UnhandledUserError extends teamsfx_api_1.UserError {
    constructor(actionName, detail, helpLink) {
        const errorOptions = generateErrorOptions(actionName, detail);
        errorOptions.helpLink = helpLink;
        super(errorOptions);
    }
}
exports.UnhandledUserError = UnhandledUserError;
class UnhandledSystemError extends teamsfx_api_1.SystemError {
    constructor(actionName, detail) {
        const errorOptions = generateErrorOptions(actionName, detail);
        super(errorOptions);
    }
}
exports.UnhandledSystemError = UnhandledSystemError;
function generateErrorOptions(actionName, detail) {
    return {
        source: actionName,
        name: errorCode,
        message: localizeUtils_1.getDefaultString(messageKey, actionName, detail),
        displayMessage: localizeUtils_1.getLocalizedString(messageKey, actionName, detail),
    };
}
