// @microsoft/teamsfx-core/build/index.js
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT license.
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CollaborationConstants = exports.FeatureId = exports.AppStudioClient = exports.getPermissionMap = exports.isExistingTabApp = exports.isValidProject = exports.environmentManager = exports.envUtil = void 0;
const tslib_1 = require("tslib");
require("reflect-metadata");
tslib_1.__exportStar(require("./core/FxCore"), exports);
tslib_1.__exportStar(require("./common/tools"), exports);
tslib_1.__exportStar(require("./common/correlator"), exports);
tslib_1.__exportStar(require("./common/local"), exports);
tslib_1.__exportStar(require("./common/deps-checker"), exports);
tslib_1.__exportStar(require("./component/debugHandler"), exports);
tslib_1.__exportStar(require("./common/samples"), exports);
tslib_1.__exportStar(require("./core/error"), exports);
tslib_1.__exportStar(require("./common/globalState"), exports);
tslib_1.__exportStar(require("./common/permissionInterface"), exports);
tslib_1.__exportStar(require("./common/featureFlags"), exports);
tslib_1.__exportStar(require("./component/migrate"), exports);
tslib_1.__exportStar(require("./common/projectSettingsHelperV3"), exports);
tslib_1.__exportStar(require("./component/constants"), exports);
tslib_1.__exportStar(require("./component/resource/appManifest/utils/utils"), exports);
tslib_1.__exportStar(require("./component/resource/azureSql/constants"), exports);
var envUtil_1 = require("./component/utils/envUtil");
Object.defineProperty(exports, "envUtil", { enumerable: true, get: function () { return envUtil_1.envUtil; } });
var environment_1 = require("./core/environment");
Object.defineProperty(exports, "environmentManager", { enumerable: true, get: function () { return environment_1.environmentManager; } });
var projectSettingsHelper_1 = require("./common/projectSettingsHelper");
Object.defineProperty(exports, "isValidProject", { enumerable: true, get: function () { return projectSettingsHelper_1.isValidProject; } });
Object.defineProperty(exports, "isExistingTabApp", { enumerable: true, get: function () { return projectSettingsHelper_1.isExistingTabApp; } });
var index_1 = require("./component/resource/aadApp/permissions/index");
Object.defineProperty(exports, "getPermissionMap", { enumerable: true, get: function () { return index_1.getPermissionMap; } });
var appStudioClient_1 = require("./component/resource/appManifest/appStudioClient");
Object.defineProperty(exports, "AppStudioClient", { enumerable: true, get: function () { return appStudioClient_1.AppStudioClient; } });
var question_1 = require("./component/question");
Object.defineProperty(exports, "FeatureId", { enumerable: true, get: function () { return question_1.FeatureId; } });
var collaborator_1 = require("./core/collaborator");
Object.defineProperty(exports, "CollaborationConstants", { enumerable: true, get: function () { return collaborator_1.CollaborationConstants; } });
