// @microsoft/teamsfx-core/build/common/m365/packageService.js
"use strict";
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT license.
Object.defineProperty(exports, "__esModule", { value: true });
exports.PackageService = void 0;
const tslib_1 = require("tslib");
const axios_1 = tslib_1.__importDefault(require("axios"));
const form_data_1 = tslib_1.__importDefault(require("form-data"));
const fs_extra_1 = tslib_1.__importDefault(require("fs-extra"));
const teamsfx_api_1 = require("@microsoft/teamsfx-api");
const tools_1 = require("../tools");
const error_1 = require("../../core/error");
// Call m365 service for package CRUD
class PackageService {
    constructor(endpoint, logger) {
        this.axiosInstance = axios_1.default.create({
            timeout: 30000,
        });
        this.initEndpoint = endpoint;
        this.logger = logger;
    }
    async getTitleServiceUrl(token) {
        var _a, _b;
        try {
            const envInfo = await this.axiosInstance.get("/config/v1/environment", {
                baseURL: this.initEndpoint,
                headers: {
                    Authorization: `Bearer ${token}`,
                },
            });
            (_a = this.logger) === null || _a === void 0 ? void 0 : _a.debug(JSON.stringify(envInfo.data));
            return envInfo.data.titlesServiceUrl;
        }
        catch (error) {
            (_b = this.logger) === null || _b === void 0 ? void 0 : _b.error(`Get ServiceUrl failed. ${error.message}`);
            throw error;
        }
    }
    async sideLoading(token, manifestPath) {
        var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k, _l;
        try {
            const data = await fs_extra_1.default.readFile(manifestPath);
            const content = new form_data_1.default();
            content.append("package", data);
            const serviceUrl = await this.getTitleServiceUrl(token);
            (_a = this.logger) === null || _a === void 0 ? void 0 : _a.info("Uploading package ...");
            const uploadHeaders = content.getHeaders();
            uploadHeaders["Authorization"] = `Bearer ${token}`;
            const uploadResponse = await this.axiosInstance.post("/dev/v1/users/packages", content.getBuffer(), {
                baseURL: serviceUrl,
                headers: uploadHeaders,
            });
            const operationId = uploadResponse.data.operationId;
            const titleId = uploadResponse.data.titlePreview.titleId;
            (_b = this.logger) === null || _b === void 0 ? void 0 : _b.debug(`Package uploaded. OperationId: ${operationId}, TitleId: ${titleId}`);
            (_c = this.logger) === null || _c === void 0 ? void 0 : _c.info("Acquiring package ...");
            const acquireResponse = await this.axiosInstance.post("/dev/v1/users/packages/acquisitions", {
                operationId: operationId,
            }, {
                baseURL: serviceUrl,
                headers: {
                    Authorization: `Bearer ${token}`,
                },
            });
            const statusId = acquireResponse.data.statusId;
            (_d = this.logger) === null || _d === void 0 ? void 0 : _d.debug(`Acquiring package with statusId: ${statusId} ...`);
            let complete = false;
            do {
                const statusResponse = await this.axiosInstance.get(`/dev/v1/users/packages/status/${statusId}`, {
                    baseURL: serviceUrl,
                    headers: { Authorization: `Bearer ${token}` },
                });
                const resCode = statusResponse.status;
                if (resCode === 200) {
                    complete = true;
                }
                else {
                    await tools_1.waitSeconds(2);
                }
            } while (complete === false);
            (_e = this.logger) === null || _e === void 0 ? void 0 : _e.info(`Acquire done. App TitleId: ${titleId}`);
            (_f = this.logger) === null || _f === void 0 ? void 0 : _f.info("Checking acquired package ...");
            const launchInfo = await this.axiosInstance.get(`/catalog/v1/users/titles/${titleId}/launchInfo`, {
                baseURL: serviceUrl,
                params: {
                    SupportedElementTypes: 
                    // eslint-disable-next-line no-secrets/no-secrets
                    "Extension,OfficeAddIn,ExchangeAddIn,FirstPartyPages,Dynamics,AAD,LineOfBusiness,LaunchPage,MessageExtension,Bot",
                },
                headers: {
                    Authorization: `Bearer ${token}`,
                },
            });
            (_g = this.logger) === null || _g === void 0 ? void 0 : _g.debug(JSON.stringify(launchInfo.data));
            (_h = this.logger) === null || _h === void 0 ? void 0 : _h.info("Sideloading done.");
            return titleId;
        }
        catch (error) {
            (_j = this.logger) === null || _j === void 0 ? void 0 : _j.error("Sideloading failed.");
            if (error.response) {
                (_k = this.logger) === null || _k === void 0 ? void 0 : _k.error(JSON.stringify(error.response.data));
            }
            else {
                (_l = this.logger) === null || _l === void 0 ? void 0 : _l.error(error.message);
            }
            throw teamsfx_api_1.assembleError(error, error_1.CoreSource);
        }
    }
    async retrieveTitleId(token, manifestId) {
        var _a, _b, _c, _d, _e, _f, _g, _h, _j;
        try {
            const serviceUrl = await this.getTitleServiceUrl(token);
            (_a = this.logger) === null || _a === void 0 ? void 0 : _a.info("Retrieve TitleId ...");
            const launchInfo = await this.axiosInstance.post("/catalog/v1/users/titles/launchInfo", {
                Id: manifestId,
                IdType: "ManifestId",
                Filter: {
                    SupportedElementTypes: [
                        // "Extensions", // Extensions require ClientDetails to be determined later
                        "OfficeAddIns",
                        "ExchangeAddIns",
                        "FirstPartyPages",
                        "Dynamics",
                        "AAD",
                        "LineOfBusiness",
                        "StaticTabs",
                        "ComposeExtensions",
                        "Bots",
                        "GraphConnector",
                        "ConfigurableTabs",
                        "Activities",
                        "MeetingExtensionDefinition",
                    ],
                },
            }, {
                baseURL: serviceUrl,
                headers: {
                    Authorization: `Bearer ${token}`,
                },
            });
            const titleId = (_d = (_c = (_b = launchInfo.data.acquisition) === null || _b === void 0 ? void 0 : _b.titleId) === null || _c === void 0 ? void 0 : _c.id) !== null && _d !== void 0 ? _d : (_e = launchInfo.data.acquisition) === null || _e === void 0 ? void 0 : _e.titleId;
            (_f = this.logger) === null || _f === void 0 ? void 0 : _f.debug(`TitleId: ${titleId}`);
            return titleId;
        }
        catch (error) {
            (_g = this.logger) === null || _g === void 0 ? void 0 : _g.error("Retrieve TitleId failed.");
            if (error.response) {
                (_h = this.logger) === null || _h === void 0 ? void 0 : _h.error(JSON.stringify(error.response.data));
            }
            else {
                (_j = this.logger) === null || _j === void 0 ? void 0 : _j.error(error.message);
            }
            throw teamsfx_api_1.assembleError(error, error_1.CoreSource);
        }
    }
    async unacquire(token, titleId) {
        var _a, _b, _c, _d, _e;
        try {
            const serviceUrl = await this.getTitleServiceUrl(token);
            (_a = this.logger) === null || _a === void 0 ? void 0 : _a.info(`Unacquiring package with TitleId ${titleId} ...`);
            await this.axiosInstance.delete(`/catalog/v1/users/acquisitions/${titleId}`, {
                baseURL: serviceUrl,
                headers: {
                    Authorization: `Bearer ${token}`,
                },
            });
            (_b = this.logger) === null || _b === void 0 ? void 0 : _b.info("Unacquiring done.");
        }
        catch (error) {
            (_c = this.logger) === null || _c === void 0 ? void 0 : _c.error("Unacquire failed.");
            if (error.response) {
                (_d = this.logger) === null || _d === void 0 ? void 0 : _d.error(JSON.stringify(error.response.data));
            }
            else {
                (_e = this.logger) === null || _e === void 0 ? void 0 : _e.error(error.message);
            }
            throw teamsfx_api_1.assembleError(error, error_1.CoreSource);
        }
    }
    async getLaunchInfo(token, titleId) {
        var _a, _b, _c, _d, _e;
        try {
            const serviceUrl = await this.getTitleServiceUrl(token);
            (_a = this.logger) === null || _a === void 0 ? void 0 : _a.info(`Getting LaunchInfo with TitleId ${titleId} ...`);
            const launchInfo = await this.axiosInstance.get(`/catalog/v1/users/titles/${titleId}/launchInfo`, {
                baseURL: serviceUrl,
                params: {
                    SupportedElementTypes: 
                    // eslint-disable-next-line no-secrets/no-secrets
                    "Extensions,OfficeAddIns,ExchangeAddIns,FirstPartyPages,Dynamics,AAD,LineOfBusiness,StaticTabs,ComposeExtensions,Bots,GraphConnector,ConfigurableTabs,Activities,MeetingExtensionDefinition",
                },
                headers: {
                    Authorization: `Bearer ${token}`,
                },
            });
            (_b = this.logger) === null || _b === void 0 ? void 0 : _b.info(JSON.stringify(launchInfo.data));
            return launchInfo.data;
        }
        catch (error) {
            (_c = this.logger) === null || _c === void 0 ? void 0 : _c.error("Get LaunchInfo failed.");
            if (error.response) {
                (_d = this.logger) === null || _d === void 0 ? void 0 : _d.error(JSON.stringify(error.response.data));
            }
            else {
                (_e = this.logger) === null || _e === void 0 ? void 0 : _e.error(error.message);
            }
            throw teamsfx_api_1.assembleError(error, error_1.CoreSource);
        }
    }
}
exports.PackageService = PackageService;
