// @microsoft/teamsfx-core/build/component/resource/botService/result.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.FxBotPluginResultFactory = void 0;
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT license.
const teamsfx_api_1 = require("@microsoft/teamsfx-api");
const constants_1 = require("./constants");
class FxBotPluginResultFactory {
    static UserError(errorName, errorMessage, innerError, helpLink) {
        return teamsfx_api_1.err(new teamsfx_api_1.UserError({
            name: errorName,
            message: errorMessage[0],
            displayMessage: errorMessage[1],
            source: FxBotPluginResultFactory.source,
            error: innerError,
            helpLink: helpLink !== null && helpLink !== void 0 ? helpLink : this.defaultHelpLink,
        }));
    }
    static SystemError(errorName, errorMessage, innerError) {
        return teamsfx_api_1.err(new teamsfx_api_1.SystemError({
            name: errorName,
            message: errorMessage[0],
            displayMessage: errorMessage[1],
            source: FxBotPluginResultFactory.source,
            error: innerError,
            issueLink: FxBotPluginResultFactory.defaultIssueLink,
        }));
    }
}
exports.FxBotPluginResultFactory = FxBotPluginResultFactory;
FxBotPluginResultFactory.source = constants_1.Alias.TEAMS_BOT_PLUGIN;
FxBotPluginResultFactory.defaultHelpLink = constants_1.Links.HELP_LINK;
FxBotPluginResultFactory.defaultIssueLink = constants_1.Links.ISSUE_LINK;
