// @microsoft/teamsfx-core/build/component/resource/spfx/utils/constants.js
"use strict";
// Copyright (c) Microsoft Corporation.
Object.defineProperty(exports, "__esModule", { value: true });
exports.UserTasks = exports.ManifestTemplate = exports.ScaffoldProgressMessage = exports.DeployProgressMessage = exports.PreDeployProgressMessage = exports.ProgressTitleMessage = exports.TelemetryEvent = exports.TelemetryValue = exports.TelemetryKey = exports.PlaceHolders = exports.Constants = void 0;
const localizeUtils_1 = require("../../../../common/localizeUtils");
// Licensed under the MIT license.
class Constants {
}
exports.Constants = Constants;
Constants.FRAMEWORK_NONE = "none";
Constants.FRAMEWORK_REACT = "react";
Constants.MAX_ALIAS_LENGTH = 40;
Constants.MAX_BUNDLE_NAME_LENGTH = 64;
Constants.CALLED_ID = "teamsdev";
Constants.APP_CATALOG_REFRESH_TIME = 20000;
Constants.APP_CATALOG_MAX_TIMES = 6;
Constants.APP_CATALOG_ACTIVE_TIME = 180000;
Constants.PLUGIN_NAME = "SPFx";
Constants.PLUGIN_DEV_NAME = "fx-resource-spfx";
Constants.BUILD_SHAREPOINT_PACKAGE = "Build SharePoint Package";
Constants.READ_MORE = localizeUtils_1.getLocalizedString("core.Notification.ReadMore");
Constants.CANCEL = localizeUtils_1.getLocalizedString("core.option.cancel");
Constants.DEPLOY_GUIDE = "https://docs.microsoft.com/en-us/microsoftteams/platform/get-started/first-app-spfx?tabs=vscode#deploy-your-app-to-sharepoint";
Constants.CREATE_APP_CATALOG_GUIDE = "https://docs.microsoft.com/en-us/sharepoint/use-app-catalog#create-the-app-catalog";
Constants.SPFX_VERSION = "1.16.1";
Constants.SUPPORTED_NODE_VERSION = ["16"];
Constants.SUPPORTED_NPM_VERSION = ["7", "8"];
Constants.SPFX_HELP_LINK = "https://aka.ms/teamsfx-spfx-help";
Constants.TEMPLATE_NAME = "spfx-tab";
class PlaceHolders {
}
exports.PlaceHolders = PlaceHolders;
PlaceHolders.componentName = "<%= componentName %>";
PlaceHolders.componentNameCamelCase = "<%= componentNameCamelCase %>";
PlaceHolders.componentClassName = "<%= componentClassName %>";
PlaceHolders.componentStrings = "<%= componentStrings %>";
PlaceHolders.libraryName = "<%= libraryName %>";
PlaceHolders.componentId = "<%= componentId %>";
PlaceHolders.componentAlias = "<%= componentAlias %>";
PlaceHolders.componentDescription = "<%= componentDescription %>";
PlaceHolders.componentNameUnescaped = "<%= componentNameUnescaped %>";
PlaceHolders.componentClassNameKebabCase = "<%= componentClassNameKebabCase %>";
class TelemetryKey {
}
exports.TelemetryKey = TelemetryKey;
TelemetryKey.Component = "component";
TelemetryKey.Success = "success";
TelemetryKey.ErrorType = "error-type";
TelemetryKey.ErrorMessage = "error-message";
TelemetryKey.ErrorCode = "error-code";
class TelemetryValue {
}
exports.TelemetryValue = TelemetryValue;
TelemetryValue.Success = "yes";
TelemetryValue.Fail = "no";
TelemetryValue.UserError = "user";
TelemetryValue.SystemError = "system";
class TelemetryEvent {
}
exports.TelemetryEvent = TelemetryEvent;
TelemetryEvent.StartSuffix = "-start";
TelemetryEvent.Scaffold = "scaffold";
TelemetryEvent.PreDeploy = "pre-deploy";
TelemetryEvent.Deploy = "deploy";
class ProgressTitleMessage {
}
exports.ProgressTitleMessage = ProgressTitleMessage;
ProgressTitleMessage.PreDeployProgressTitle = localizeUtils_1.getLocalizedString("plugins.spfx.buildSharepointPackage");
ProgressTitleMessage.DeployProgressTitle = localizeUtils_1.getLocalizedString("plugins.spfx.deploy.title");
ProgressTitleMessage.ScaffoldProgressTitle = localizeUtils_1.getLocalizedString("plugins.spfx.scaffold.title");
class PreDeployProgressMessage {
}
exports.PreDeployProgressMessage = PreDeployProgressMessage;
PreDeployProgressMessage.NpmInstall = "Run: npm install. This may take more than 5 minutes to finish";
PreDeployProgressMessage.GulpBundle = "Run: gulp bundle --ship";
PreDeployProgressMessage.GulpPackage = "Run: gulp package-solution --ship";
class DeployProgressMessage {
}
exports.DeployProgressMessage = DeployProgressMessage;
DeployProgressMessage.CreateSPAppCatalog = localizeUtils_1.getLocalizedString("plugins.spfx.deploy.createAppcatalog");
DeployProgressMessage.UploadAndDeploy = localizeUtils_1.getLocalizedString("plugins.spfx.deploy.uploadAddDeploy");
class ScaffoldProgressMessage {
}
exports.ScaffoldProgressMessage = ScaffoldProgressMessage;
ScaffoldProgressMessage.DependencyCheck = localizeUtils_1.getLocalizedString("plugins.spfx.scaffold.dependencyCheck");
ScaffoldProgressMessage.DependencyInstall = localizeUtils_1.getLocalizedString("plugins.spfx.scaffold.dependencyInstall");
ScaffoldProgressMessage.ScaffoldProject = localizeUtils_1.getLocalizedString("plugins.spfx.scaffold.scaffoldProject");
ScaffoldProgressMessage.UpdateManifest = localizeUtils_1.getLocalizedString("plugins.spfx.scaffold.updateManifest");
class ManifestTemplate {
}
exports.ManifestTemplate = ManifestTemplate;
ManifestTemplate.LOCAL_CONTENT_URL = "https://{teamSiteDomain}/_layouts/15/TeamsLogon.aspx?SPFX=true&dest={teamSitePath}/_layouts/15/TeamsWorkBench.aspx%3FcomponentId=%s%26teams%26personal%26forceLocale={locale}%26loadSPFX%3Dtrue%26debugManifestsFile%3Dhttps%3A%2F%2Flocalhost%3A4321%2Ftemp%2Fmanifests.js";
ManifestTemplate.LOCAL_CONFIGURATION_URL = "https://{teamSiteDomain}{teamSitePath}/_layouts/15/TeamsLogon.aspx?SPFX=true&dest={teamSitePath}/_layouts/15/TeamsWorkBench.aspx%3FcomponentId=%s%26openPropertyPane=true%26teams%26forceLocale={locale}%26loadSPFX%3Dtrue%26debugManifestsFile%3Dhttps%3A%2F%2Flocalhost%3A4321%2Ftemp%2Fmanifests.js";
ManifestTemplate.REMOTE_CONTENT_URL = "{{^config.isLocalDebug}}https://{teamSiteDomain}/_layouts/15/TeamsLogon.aspx?SPFX=true&dest=/_layouts/15/teamshostedapp.aspx%3Fteams%26personal%26componentId=%s%26forceLocale={locale}{{/config.isLocalDebug}}{{#config.isLocalDebug}}https://{teamSiteDomain}/_layouts/15/TeamsLogon.aspx?SPFX=true&dest={teamSitePath}/_layouts/15/TeamsWorkBench.aspx%3FcomponentId=%s%26teams%26personal%26forceLocale={locale}%26loadSPFX%3Dtrue%26debugManifestsFile%3Dhttps%3A%2F%2Flocalhost%3A4321%2Ftemp%2Fmanifests.js{{/config.isLocalDebug}}";
ManifestTemplate.REMOTE_CONFIGURATION_URL = "{{^config.isLocalDebug}}https://{teamSiteDomain}{teamSitePath}/_layouts/15/TeamsLogon.aspx?SPFX=true&dest={teamSitePath}/_layouts/15/teamshostedapp.aspx%3FopenPropertyPane=true%26teams%26componentId=%s%26forceLocale={locale}{{/config.isLocalDebug}}{{#config.isLocalDebug}}https://{teamSiteDomain}{teamSitePath}/_layouts/15/TeamsLogon.aspx?SPFX=true&dest={teamSitePath}/_layouts/15/TeamsWorkBench.aspx%3FcomponentId=%s%26openPropertyPane=true%26teams%26forceLocale={locale}%26loadSPFX%3Dtrue%26debugManifestsFile%3Dhttps%3A%2F%2Flocalhost%3A4321%2Ftemp%2Fmanifests.js{{/config.isLocalDebug}}";
ManifestTemplate.WEBSITE_URL = "https://products.office.com/en-us/sharepoint/collaboration";
ManifestTemplate.WEB_APP_INFO_RESOURCE = "https://{teamSiteDomain}";
ManifestTemplate.WEB_APP_INFO_ID = "00000003-0000-0ff1-ce00-000000000000";
var UserTasks;
(function (UserTasks) {
    UserTasks["AddFeature"] = "addFeature";
})(UserTasks = exports.UserTasks || (exports.UserTasks = {}));
