// @microsoft/teamsfx-core/build/component/resource/spfx/utils/progress-helper.js
"use strict";
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT license.
Object.defineProperty(exports, "__esModule", { value: true });
exports.ProgressHelper = void 0;
const constants_1 = require("./constants");
class ProgressHelper {
    static async startPreDeployProgressHandler(ui) {
        var _a;
        this.preDeployProgress = ui === null || ui === void 0 ? void 0 : ui.createProgressBar(constants_1.ProgressTitleMessage.PreDeployProgressTitle, Object.entries(constants_1.PreDeployProgressMessage).length);
        await ((_a = this.preDeployProgress) === null || _a === void 0 ? void 0 : _a.start(""));
        return this.preDeployProgress;
    }
    static async endAllHandlers(success) {
        await this.endPreDeployProgress(success);
    }
    static async endPreDeployProgress(success) {
        var _a;
        await ((_a = this.preDeployProgress) === null || _a === void 0 ? void 0 : _a.end(success));
        this.preDeployProgress = undefined;
    }
    static async startDeployProgressHandler(ui) {
        var _a;
        this.deployProgress = ui === null || ui === void 0 ? void 0 : ui.createProgressBar(constants_1.ProgressTitleMessage.DeployProgressTitle, Object.entries(constants_1.DeployProgressMessage).length);
        await ((_a = this.deployProgress) === null || _a === void 0 ? void 0 : _a.start(""));
        return this.deployProgress;
    }
    static async endDeployProgress(success) {
        var _a;
        await ((_a = this.deployProgress) === null || _a === void 0 ? void 0 : _a.end(success));
        this.deployProgress = undefined;
    }
    static async startScaffoldProgressHandler(ui) {
        var _a;
        this.scaffoldProgress = ui === null || ui === void 0 ? void 0 : ui.createProgressBar(constants_1.ProgressTitleMessage.ScaffoldProgressTitle, Object.entries(constants_1.ScaffoldProgressMessage).length);
        await ((_a = this.scaffoldProgress) === null || _a === void 0 ? void 0 : _a.start(""));
        return this.scaffoldProgress;
    }
    static async endScaffoldProgress(success) {
        var _a;
        await ((_a = this.scaffoldProgress) === null || _a === void 0 ? void 0 : _a.end(success));
        this.scaffoldProgress = undefined;
    }
}
exports.ProgressHelper = ProgressHelper;
