// @microsoft/teamsfx-core/build/component/resource/keyVault.js
"use strict";
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT license.
Object.defineProperty(exports, "__esModule", { value: true });
exports.KeyVaultResource = void 0;
const tslib_1 = require("tslib");
const lib_1 = require("@feathersjs/hooks/lib");
require("reflect-metadata");
const typedi_1 = require("typedi");
const constants_1 = require("../constants");
const actionExecutionMW_1 = require("../middleware/actionExecutionMW");
const azureResource_1 = require("./azureResource");
let KeyVaultResource = class KeyVaultResource extends azureResource_1.AzureResource {
    constructor() {
        super(...arguments);
        this.name = constants_1.ComponentNames.KeyVault;
        this.bicepModuleName = "keyVault";
        this.outputs = constants_1.KeyVaultOutputs;
        this.finalOutputKeys = ["keyVaultResourceId", "m365ClientSecretReference", "botClientSecretReference"];
        this.templateContext = {
            identity: {
                principalId: constants_1.IdentityOutputs.identityPrincipalId.bicepVariable,
            },
        };
    }
    async generateBicep(context, inputs) {
        return super.generateBicep(context, inputs);
    }
};
tslib_1.__decorate([
    lib_1.hooks([
        actionExecutionMW_1.ActionExecutionMW({
            enableTelemetry: true,
            telemetryComponentName: "fx-resource-key-vault",
            telemetryEventName: "generate-arm-templates",
            errorSource: "kv",
        }),
    ]),
    tslib_1.__metadata("design:type", Function),
    tslib_1.__metadata("design:paramtypes", [Object, Object]),
    tslib_1.__metadata("design:returntype", Promise)
], KeyVaultResource.prototype, "generateBicep", null);
KeyVaultResource = tslib_1.__decorate([
    typedi_1.Service(constants_1.ComponentNames.KeyVault)
], KeyVaultResource);
exports.KeyVaultResource = KeyVaultResource;
