// @microsoft/teamsfx-core/build/component/resource/azureAppService/azureWebApp.js
"use strict";
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT license.
Object.defineProperty(exports, "__esModule", { value: true });
exports.AzureWebAppResource = void 0;
const tslib_1 = require("tslib");
const typedi_1 = require("typedi");
const constants_1 = require("../../constants");
const azureAppService_1 = require("./azureAppService");
let AzureWebAppResource = class AzureWebAppResource extends azureAppService_1.AzureAppService {
    constructor() {
        super(...arguments);
        this.name = "azure-web-app";
        this.alias = "WebApp";
        this.displayName = "Azure Web App";
        this.bicepModuleName = "azureWebApp";
        this.outputs = constants_1.WebAppOutputs;
        this.finalOutputKeys = ["resourceId", "endpoint"];
        this.templateContext = {
            identity: {
                resourceId: constants_1.IdentityOutputs.identityResourceId.bicepVariable,
            },
        };
    }
    async deploy(context, inputs) {
        let resourceIdKey = this.outputs.resourceId.key;
        const state = context.envInfo.state[inputs.componentId];
        if (!state[resourceIdKey]) {
            if (state["botWebAppResourceId"]) {
                resourceIdKey = "botWebAppResourceId";
            }
            if (state["webAppResourceId"]) {
                resourceIdKey = "webAppResourceId";
            }
        }
        return await super.deploy(context, inputs, false, resourceIdKey);
    }
};
AzureWebAppResource = tslib_1.__decorate([
    typedi_1.Service("azure-web-app")
], AzureWebAppResource);
exports.AzureWebAppResource = AzureWebAppResource;
