// @microsoft/teamsfx-core/build/component/resource/azureSql/constants.js
"use strict";
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT license.
Object.defineProperty(exports, "__esModule", { value: true });
exports.Message = exports.AzureSqlBicepFile = exports.AzureSqlBicep = exports.HelpLinks = exports.Telemetry = exports.ActionGenerateBicep = exports.ActionConfigure = exports.ActionProvision = exports.Constants = void 0;
class Constants {
}
exports.Constants = Constants;
Constants.pluginName = "SQL Plugin";
Constants.pluginNameShort = "SQL";
Constants.pluginFullName = "fx-resource-azure-sql";
Constants.firewall = {
    localRule: "AllowLocal",
};
Constants.ipBeginToken = "0";
Constants.ipEndToken = "255";
Constants.maxRetryTimes = 3;
Constants.jwtToken = {
    ver1: "1.0",
    ver2: "2.0",
    userType: "0",
};
Constants.sqlAdministratorType = "ActiveDirectory";
Constants.echoIpAddress = "https://api.ipify.org";
Constants.azureSqlScope = "https://database.windows.net//.default";
Constants.resourceGroupName = "resourceGroupName";
Constants.sqlEndpoint = "sqlEndpoint";
Constants.sqlResourceId = "sqlResourceId";
Constants.databaseName = "databaseName";
Constants.skipAddingSqlUser = "skipAddingSqlUser";
Constants.admin = "admin";
Constants.adminPassword = "adminPassword";
Constants.solution = "solution";
Constants.solutionPluginFullName = "fx-solution-azure";
Constants.solutionConfigKey = {
    resourceNameSuffix: "resourceNameSuffix",
    location: "location",
    tenantId: "tenantId",
};
Constants.identityPlugin = "fx-resource-identity";
Constants.identityName = "identityName";
Constants.questionKey = {
    adminName: "sql-admin-name",
    adminPassword: "sql-password",
    confirmPassword: "sql-confirm-password",
    skipAddingUser: "sql-skip-adding-user",
};
exports.ActionProvision = {
    name: "provision",
};
exports.ActionConfigure = {
    name: "configure",
};
exports.ActionGenerateBicep = {
    name: "generateBicep",
};
class Telemetry {
}
exports.Telemetry = Telemetry;
Telemetry.componentName = "fx-resource-azure-sql";
Telemetry.stage = {
    preProvision: "pre-provision",
    postProvision: "post-provision",
    getQuestion: "get-question",
    generateArmTemplates: "generate-arm-templates",
};
Telemetry.properties = {
    skipAddingUser: "skip-adding-user",
    dbCount: "db-count",
    dbOnly: "db-only",
};
class HelpLinks {
}
exports.HelpLinks = HelpLinks;
HelpLinks.default = "https://aka.ms/teamsfx-sql-help";
class AzureSqlBicep {
}
exports.AzureSqlBicep = AzureSqlBicep;
AzureSqlBicep.sqlResourceId = "provisionOutputs.azureSqlOutput.value.sqlResourceId";
AzureSqlBicep.sqlEndpoint = "provisionOutputs.azureSqlOutput.value.sqlEndpoint";
AzureSqlBicep.databaseName = "provisionOutputs.azureSqlOutput.value.databaseName";
class AzureSqlBicepFile {
}
exports.AzureSqlBicepFile = AzureSqlBicepFile;
AzureSqlBicepFile.moduleTemplateFileName = "provision.template.bicep";
AzureSqlBicepFile.ProvisionModuleTemplateFileName = "sqlProvision.template.bicep";
AzureSqlBicepFile.newDatabaseOrchestrationTemplateFileName = "newDatabase.orchestration.template.bicep";
AzureSqlBicepFile.newDatabaseProvisionTemplateFileName = "newDatabaseProvision.template.bicep";
exports.Message = {
    skipAddAadAdmin: `skip adding existing aad admin`,
    addFirewall: `add firewall`,
    addSqlAadAdmin: `add SQL aad admin`,
    addDatabaseUser: (name) => `add database user ${name}`,
};
