// @microsoft/teamsfx-core/build/component/resource/azureSql/results.js
"use strict";
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT license.
Object.defineProperty(exports, "__esModule", { value: true });
exports.SqlResultFactory = void 0;
const teamsfx_api_1 = require("@microsoft/teamsfx-api");
const constants_1 = require("./constants");
class SqlResultFactory {
    static UserError(name, messages, innerError, stack, helpLink) {
        return new teamsfx_api_1.UserError({
            name,
            message: messages[0],
            displayMessage: messages[1],
            source: this.source,
            helpLink,
            error: innerError,
        });
    }
    static SystemError(name, messages, innerError, stack, issueLink) {
        return new teamsfx_api_1.SystemError({
            name,
            message: messages[0],
            displayMessage: messages[1],
            source: this.source,
            issueLink,
            error: innerError,
        });
    }
}
exports.SqlResultFactory = SqlResultFactory;
SqlResultFactory.source = constants_1.Constants.pluginNameShort;
SqlResultFactory.defaultHelpLink = "";
SqlResultFactory.defaultIssueLink = "";
