// @microsoft/teamsfx-core/build/common/localStateConstants.js
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT license.
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.LocalStateBotKeys = exports.LocalStateBackendKeys = exports.LocalStateFrontendKeys = exports.LocalStateSimpleAuthKeys = exports.LocalStateAuthKeys = exports.LocalStateTeamsAppKeys = void 0;
exports.LocalStateTeamsAppKeys = Object.freeze({
    TeamsAppId: "teamsAppId",
});
exports.LocalStateAuthKeys = Object.freeze({
    ClientId: "clientId",
    ClientSecret: "clientSecret",
    ObjectId: "objectId",
    Oauth2PermissionScopeId: "oauth2PermissionScopeId",
    OauthAuthority: "oauthAuthority",
    OauthHost: "oauthHost",
    ApplicationIdUris: "applicationIdUris",
    TenantId: "tenantId",
});
exports.LocalStateSimpleAuthKeys = Object.freeze({
    SimpleAuthFilePath: "simpleAuthFilePath",
    EnvironmentVariableParams: "environmentVariableParams",
    Endpoint: "endpoint",
});
exports.LocalStateFrontendKeys = Object.freeze({
    TabDomain: "tabDomain",
    TabEndpoint: "tabEndpoint",
    TabIndexPath: "tabIndexPath",
    Browser: "browser",
    Https: "https",
    SslCertFile: "sslCertFile",
    SslKeyFile: "sslKeyFile",
    Endpoint: "endpoint",
});
exports.LocalStateBackendKeys = Object.freeze({
    FunctionEndpoint: "functionEndpoint",
});
exports.LocalStateBotKeys = Object.freeze({
    BotId: "botId",
    BotPassword: "botPassword",
    BotAadObjectId: "botAadObjectId",
    BotDomain: "botDomain",
    BotEndpoint: "botEndpoint",
});
