// @microsoft/teamsfx-core/build/common/localSettingsConstants.js
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT license.
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.LocalSettingsEncryptKeys = exports.LocalSettingsBotKeys = exports.LocalSettingsBackendKeys = exports.LocalSettingsFrontendKeys = exports.LocalSettingsSimpleAuthKeys = exports.LocalSettingsAuthKeys = exports.LocalSettingsTeamsAppKeys = void 0;
exports.LocalSettingsTeamsAppKeys = Object.freeze({
    TenantId: "tenantId",
    TeamsAppId: "teamsAppId",
});
exports.LocalSettingsAuthKeys = Object.freeze({
    ClientId: "clientId",
    ClientSecret: "clientSecret",
    ObjectId: "objectId",
    Oauth2PermissionScopeId: "oauth2PermissionScopeId",
    OauthAuthority: "oauthAuthority",
    OauthHost: "oauthHost",
    ApplicationIdUris: "applicationIdUris",
});
exports.LocalSettingsSimpleAuthKeys = Object.freeze({
    SimpleAuthFilePath: "simpleAuthFilePath",
    SimpleAuthEnvironmentVariableParams: "SimpleAuthEnvironmentVariableParams",
    SimpleAuthServiceEndpoint: "AuthServiceEndpoint",
});
exports.LocalSettingsFrontendKeys = Object.freeze({
    TabDomain: "tabDomain",
    TabEndpoint: "tabEndpoint",
    TabIndexPath: "tabIndexPath",
    Browser: "browser",
    Https: "https",
    SslCertFile: "sslCertFile",
    SslKeyFile: "sslKeyFile",
});
exports.LocalSettingsBackendKeys = Object.freeze({
    FunctionEndpoint: "functionEndpoint",
    FunctionName: "functionName",
});
exports.LocalSettingsBotKeys = Object.freeze({
    BotId: "botId",
    BotPassword: "botPassword",
    BotAadObjectId: "botAadObjectId",
    BotRedirectUri: "botRedirectUri",
    BotDomain: "botDomain",
    BotEndpoint: "botEndpoint",
});
exports.LocalSettingsEncryptKeys = Object.freeze({
    ClientSecret: "clientSecret",
    SimpleAuthEnvironmentVariableParams: "SimpleAuthEnvironmentVariableParams",
    BotPassword: "botPassword",
});
