// @microsoft/teamsfx-core/build/common/projectSettingsHelper.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.isVSProject = exports.newProjectSettings = exports.getProjectSettingsVersion = exports.isExistingTabApp = exports.getAzurePlugins = exports.hasAzureResource = exports.hasSPFx = exports.hasAAD = exports.isValidProjectV2 = exports.isValidProjectV3 = exports.isValidProject = exports.validateProjectSettings = void 0;
const tslib_1 = require("tslib");
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT license.
const teamsfx_api_1 = require("@microsoft/teamsfx-api");
const fs_extra_1 = tslib_1.__importDefault(require("fs-extra"));
const path = tslib_1.__importStar(require("path"));
const constants_1 = require("../component/constants");
const uuid = tslib_1.__importStar(require("uuid"));
const tools_1 = require("./tools");
const versionMetadata_1 = require("./versionMetadata");
function validateProjectSettings(projectSettings) {
    var _a;
    if (!projectSettings)
        return "empty projectSettings";
    if (!projectSettings.solutionSettings)
        return undefined;
    const solutionSettings = projectSettings.solutionSettings;
    let validateRes = validateStringArray(solutionSettings.azureResources);
    if (validateRes) {
        return `solutionSettings.azureResources validation failed: ${validateRes}`;
    }
    validateRes = validateStringArray(solutionSettings.capabilities, [
        constants_1.TabOptionItem().id,
        constants_1.BotOptionItem().id,
        constants_1.MessageExtensionItem().id,
        constants_1.TabSPFxItem().id,
        constants_1.TabSsoItem().id,
        constants_1.BotSsoItem().id,
    ]);
    if (validateRes) {
        return `solutionSettings.capabilities validation failed: ${validateRes}`;
    }
    validateRes = validateStringArray(solutionSettings.activeResourcePlugins);
    if (validateRes) {
        return `solutionSettings.activeResourcePlugins validation failed: ${validateRes}`;
    }
    if ((_a = projectSettings === null || projectSettings === void 0 ? void 0 : projectSettings.solutionSettings) === null || _a === void 0 ? void 0 : _a.migrateFromV1) {
        return "The project created before v2.0.0 is only supported in the Teams Toolkit before v3.4.0.";
    }
    return undefined;
}
exports.validateProjectSettings = validateProjectSettings;
function validateStringArray(arr, enums) {
    if (!arr) {
        return "is undefined";
    }
    if (!Array.isArray(arr)) {
        return "is not array";
    }
    for (const element of arr) {
        if (typeof element !== "string") {
            return "array elements is not string type";
        }
        if (enums && !enums.includes(element)) {
            return `array elements is out of scope: ${enums}`;
        }
    }
    return undefined;
}
function isValidProject(workspacePath) {
    if (!workspacePath)
        return false;
    try {
        if (tools_1.isV3Enabled()) {
            return isValidProjectV3(workspacePath) || isValidProjectV2(workspacePath);
        }
        else {
            return isValidProjectV2(workspacePath);
        }
    }
    catch (e) {
        return false;
    }
}
exports.isValidProject = isValidProject;
function isValidProjectV3(workspacePath) {
    // TODO: should be cleaned after v3 folder changed.
    const filePath = path.resolve(workspacePath, teamsfx_api_1.SettingsFolderName, teamsfx_api_1.SettingsFileName);
    if (fs_extra_1.default.existsSync(filePath)) {
        const projectSettings = fs_extra_1.default.readJsonSync(filePath);
        if (!projectSettings.trackingId) {
            return false;
        }
        if (!projectSettings.version) {
            return false;
        }
        return true;
    }
    const ymlFilePath = path.join(workspacePath, versionMetadata_1.MetadataV3.configFile);
    const localYmlPath = path.join(workspacePath, versionMetadata_1.MetadataV3.localConfigFile);
    if (fs_extra_1.default.pathExistsSync(ymlFilePath) || fs_extra_1.default.pathExistsSync(localYmlPath)) {
        return true;
    }
    return false;
}
exports.isValidProjectV3 = isValidProjectV3;
function isValidProjectV2(workspacePath) {
    const confFolderPath = path.resolve(workspacePath, `.${teamsfx_api_1.ConfigFolderName}`, "configs");
    const settingsFile = path.resolve(confFolderPath, teamsfx_api_1.ProjectSettingsFileName);
    if (!fs_extra_1.default.existsSync(settingsFile)) {
        return false;
    }
    const projectSettings = fs_extra_1.default.readJsonSync(settingsFile);
    if (validateProjectSettings(projectSettings))
        return false;
    return true;
}
exports.isValidProjectV2 = isValidProjectV2;
function hasAAD(projectSetting) {
    const solutionSettings = projectSetting.solutionSettings;
    if (!solutionSettings)
        return false;
    return solutionSettings.activeResourcePlugins.includes(constants_1.BuiltInFeaturePluginNames.aad);
}
exports.hasAAD = hasAAD;
function hasSPFx(projectSetting) {
    const solutionSettings = projectSetting.solutionSettings;
    if (!solutionSettings)
        return false;
    return solutionSettings.activeResourcePlugins.includes(constants_1.BuiltInFeaturePluginNames.spfx);
}
exports.hasSPFx = hasSPFx;
function hasAzureResource(projectSetting, excludeAad = false) {
    const solutionSettings = projectSetting.solutionSettings;
    if (!solutionSettings)
        return false;
    const azurePlugins = getAzurePlugins(excludeAad);
    for (const pluginName of solutionSettings.activeResourcePlugins) {
        if (azurePlugins.includes(pluginName))
            return true;
    }
    return false;
}
exports.hasAzureResource = hasAzureResource;
function getAzurePlugins(excludeAad = false) {
    const azurePlugins = [
        constants_1.BuiltInFeaturePluginNames.apim,
        constants_1.BuiltInFeaturePluginNames.bot,
        constants_1.BuiltInFeaturePluginNames.frontend,
        constants_1.BuiltInFeaturePluginNames.function,
        constants_1.BuiltInFeaturePluginNames.identity,
        constants_1.BuiltInFeaturePluginNames.keyVault,
        constants_1.BuiltInFeaturePluginNames.simpleAuth,
        constants_1.BuiltInFeaturePluginNames.sql,
    ];
    if (!excludeAad) {
        azurePlugins.push(constants_1.BuiltInFeaturePluginNames.aad);
    }
    return azurePlugins;
}
exports.getAzurePlugins = getAzurePlugins;
function isExistingTabApp(projectSettings) {
    var _a;
    if (!tools_1.isExistingTabAppEnabled()) {
        return false;
    }
    const solutionSettings = projectSettings.solutionSettings;
    if (!solutionSettings) {
        return true;
    }
    // Scenario: SSO is added to existing tab app
    if (((_a = solutionSettings.capabilities) === null || _a === void 0 ? void 0 : _a.length) === 1 &&
        solutionSettings.capabilities.includes(constants_1.TabSsoItem().id)) {
        return true;
    }
    return false;
}
exports.isExistingTabApp = isExistingTabApp;
function getProjectSettingsVersion() {
    return "2.1.0";
}
exports.getProjectSettingsVersion = getProjectSettingsVersion;
function newProjectSettings() {
    const projectSettings = {
        appName: "",
        projectId: uuid.v4(),
        version: getProjectSettingsVersion(),
    };
    return projectSettings;
}
exports.newProjectSettings = newProjectSettings;
function isVSProject(projectSettings) {
    return (projectSettings === null || projectSettings === void 0 ? void 0 : projectSettings.programmingLanguage) === "csharp";
}
exports.isVSProject = isVSProject;
