// @microsoft/teamsfx-core/build/common/projectSettingsHelperV3.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.isMiniApp = exports.hasSQL = exports.hasKeyVault = exports.hasAPIM = exports.hasSPFxTab = exports.hasAzureResourceV3 = exports.hasSimpleAuth = exports.hasApi = exports.hasAAD = exports.hasFunctionBot = exports.hasBot = exports.hasAzureTab = exports.hasTab = exports.validateProjectSettings = void 0;
const constants_1 = require("../component/constants");
const workflow_1 = require("../component/workflow");
function validateProjectSettings(projectSettings) {
    if (!projectSettings)
        return "empty projectSettings";
    const components = projectSettings.components;
    if (!components)
        return "components is undefined";
    return undefined;
}
exports.validateProjectSettings = validateProjectSettings;
function hasTab(projectSettings) {
    const components = projectSettings.components;
    return components.filter((c) => c.name === constants_1.ComponentNames.TeamsTab).length > 0;
}
exports.hasTab = hasTab;
function hasAzureTab(projectSettings) {
    const tab = workflow_1.getComponent(projectSettings, constants_1.ComponentNames.TeamsTab);
    return tab !== undefined && tab.hosting !== constants_1.ComponentNames.SPFx;
}
exports.hasAzureTab = hasAzureTab;
function hasBot(projectSettings) {
    const components = projectSettings.components;
    return components.filter((c) => c.name === constants_1.ComponentNames.TeamsBot).length > 0;
}
exports.hasBot = hasBot;
function hasFunctionBot(projectSettings) {
    const botComponent = workflow_1.getComponent(projectSettings, constants_1.ComponentNames.TeamsBot);
    if (!botComponent)
        return false;
    return botComponent.hosting === constants_1.ComponentNames.Function;
}
exports.hasFunctionBot = hasFunctionBot;
function hasAAD(projectSettings) {
    const components = projectSettings.components;
    return components.filter((c) => c.name === constants_1.ComponentNames.AadApp).length > 0;
}
exports.hasAAD = hasAAD;
function hasApi(projectSettings) {
    const components = projectSettings.components;
    return components.filter((c) => c.name === constants_1.ComponentNames.TeamsApi).length > 0;
}
exports.hasApi = hasApi;
function hasSimpleAuth(projectSettings) {
    const components = projectSettings.components;
    return components.filter((c) => c.name === constants_1.ComponentNames.SimpleAuth).length > 0;
}
exports.hasSimpleAuth = hasSimpleAuth;
function hasAzureResourceV3(projectSetting, excludeAad = false) {
    var _a;
    const filtered = (_a = projectSetting.components) === null || _a === void 0 ? void 0 : _a.filter((c) => constants_1.AzureResources.includes(c.name) || (!excludeAad && c.name === constants_1.ComponentNames.AadApp));
    return filtered.length > 0;
}
exports.hasAzureResourceV3 = hasAzureResourceV3;
function hasSPFxTab(projectSetting) {
    const tab = workflow_1.getComponent(projectSetting, constants_1.ComponentNames.TeamsTab);
    return (tab === null || tab === void 0 ? void 0 : tab.hosting) === constants_1.ComponentNames.SPFx;
}
exports.hasSPFxTab = hasSPFxTab;
function hasAPIM(projectSettings) {
    return workflow_1.getComponent(projectSettings, constants_1.ComponentNames.APIM) !== undefined;
}
exports.hasAPIM = hasAPIM;
function hasKeyVault(projectSettings) {
    return workflow_1.getComponent(projectSettings, constants_1.ComponentNames.KeyVault) !== undefined;
}
exports.hasKeyVault = hasKeyVault;
function hasSQL(projectSettings) {
    return workflow_1.getComponent(projectSettings, constants_1.ComponentNames.AzureSQL) !== undefined;
}
exports.hasSQL = hasSQL;
function isMiniApp(projectSettings) {
    if (!projectSettings.components || projectSettings.components.length === 0)
        return true;
    // Scenario: SSO is added to existing tab app
    if (projectSettings.components.length === 1 && hasAAD(projectSettings)) {
        return true;
    }
    return false;
}
exports.isMiniApp = isMiniApp;
