// @microsoft/teamsfx-core/build/component/code/constants.js
"use strict";
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT license.
Object.defineProperty(exports, "__esModule", { value: true });
exports.TelemetryComponent = exports.RemoteTeamsAppId = exports.TemplatePlaceHolders = exports.ApiConstants = exports.TemplateGroup = exports.DEFAULT_DOTNET_FRAMEWORK = exports.NpmScripts = exports.Commands = void 0;
class Commands {
}
exports.Commands = Commands;
Commands.NpmInstall = "npm install";
Commands.NpmInstallProd = "npm install --only=prod";
Commands.NpmBuild = "npm run build";
Commands.DotNetPublish = "dotnet publish --configuration Release";
Commands.BlazorBuild = (output, runtime) => `dotnet publish --output ${output} --configuration Release --runtime ${runtime} --self-contained`;
Commands.NpmRunScript = (script) => `npm run ${script}`;
exports.NpmScripts = {
    customizedBuild: "build:teamsfx",
    customizedInstall: "install:teamsfx",
};
exports.DEFAULT_DOTNET_FRAMEWORK = "net6.0";
exports.TemplateGroup = {
    bot: "bot",
    tab: "tab",
    apiBase: "function-base",
    apiTriggers: "function-triggers",
};
exports.ApiConstants = {
    functionTriggerType: "HTTPTrigger",
    baseScenarioName: "default",
};
exports.TemplatePlaceHolders = {
    functionEntry: /entryname/g,
    ProjectFile: /ProjectName/g,
};
exports.RemoteTeamsAppId = "remoteTeamsAppId";
exports.TelemetryComponent = {
    api: "fx-resource-function",
};
