// @azure/arm-botservice/esm/operations/directLine.js
/*
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License. See License.txt in the project root for
 * license information.
 *
 * Code generated by Microsoft (R) AutoRest Code Generator.
 * Changes may cause incorrect behavior and will be lost if the code is
 * regenerated.
 */
import { __assign } from "tslib";
import * as msRest from "@azure/ms-rest-js";
import * as Mappers from "../models/directLineMappers";
import * as Parameters from "../models/parameters";
/** Class representing a DirectLine. */
var DirectLine = /** @class */ (function () {
    /**
     * Create a DirectLine.
     * @param {AzureBotServiceContext} client Reference to the service client.
     */
    function DirectLine(client) {
        this.client = client;
    }
    DirectLine.prototype.regenerateKeys = function (resourceGroupName, resourceName, channelName, parameters, options, callback) {
        return this.client.sendOperationRequest({
            resourceGroupName: resourceGroupName,
            resourceName: resourceName,
            channelName: channelName,
            parameters: parameters,
            options: options
        }, regenerateKeysOperationSpec, callback);
    };
    return DirectLine;
}());
export { DirectLine };
// Operation Specifications
var serializer = new msRest.Serializer(Mappers);
var regenerateKeysOperationSpec = {
    httpMethod: "POST",
    path: "subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.BotService/botServices/{resourceName}/channels/{channelName}/regeneratekeys",
    urlParameters: [
        Parameters.resourceGroupName,
        Parameters.resourceName,
        Parameters.subscriptionId,
        Parameters.channelName2
    ],
    queryParameters: [
        Parameters.apiVersion
    ],
    headerParameters: [
        Parameters.acceptLanguage
    ],
    requestBody: {
        parameterPath: "parameters",
        mapper: __assign(__assign({}, Mappers.SiteInfo), { required: true })
    },
    responses: {
        200: {
            bodyMapper: Mappers.BotChannel
        },
        default: {
            bodyMapper: Mappers.ErrorModel
        }
    },
    serializer: serializer
};
